{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "84dad7d0",
   "metadata": {},
   "source": [
    "# TWAS & Coloc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a4d1e10",
   "metadata": {},
   "source": [
    "## 1. TWAS\n",
    "### 1.1 Transcriptome-Wide Association Study (TWAS)\n",
    "\n",
    "**TWAS** is a method used to identify genes associated with complex traits by leveraging predicted gene expression levels. It combines precomputed molecular phenotype prediction weights with GWAS summary statistics to perform gene-trait association tests. \n",
    "\n",
    "\n",
    "#### 1.2 Input File Formats\n",
    "\n",
    "- GWAS Metadata (`--gwas_meta_data`)    \n",
    "A TSV file specifying study ID, chromosome, path to summary statistics, and a column mapping file.  \n",
    "- LD Reference Metadata (`--ld_meta_data`)    \n",
    "A TSV file with chromosome, start and end positions, and paths to LD matrix files.  \n",
    "- xQTL Weight Metadata (`--xqtl_meta_data`)    \n",
    "Metadata file containing gene region coordinates and paths to weight files stored in RDS format.  \n",
    "- Genomic Region Definitions (`--regions`)    \n",
    "A BED file defining analysis regions.  \n",
    "- xQTL Type Table (`--xqtl_type_table`)    \n",
    "A mapping file defining the molecular phenotype types and their associated biological context.\n",
    "\n",
    "\n",
    "#### 1.3 Key Parameters\n",
    "\n",
    "- `--rsq_cutoff 0.01`: Cross-validation RÂ² threshold used to determine whether a gene is predictable  \n",
    "- `--rsq_pval_cutoff 0.05`: Cross-validation p-value threshold  \n",
    "\n",
    "\n",
    "#### 1.4 Analysis Workflow  \n",
    "\n",
    "1). **Weight Loading**: Load TWAS prediction weights from RDS files.     \n",
    "2). **Predictable Gene Identification**: Determine which genes are reliably predictable based on cross-validation performance.    \n",
    "3). **TWAS Testing**: Compute TWAS Z-scores by combining predicted expression with GWAS statistics.    \n",
    "4). **Mendelian Randomization (MR)**: Perform MR analysis on candidate genes.    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09dbf063",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "study_id\tchrom\tfile_path\tcolumn_mapping_file\n",
      "Bellenguez_2022\t11\tdata/twas/RSS_QC_RAISS_imputed.AD_Bellenguez_2022_April9_chr11.tsv.gz\tdata/twas/Bellenguez.yml\n"
     ]
    }
   ],
   "source": [
    "# gwas input \n",
    "cd /home/ubuntu/xqtl_protocol_exercise\n",
    "head data/twas/gwas_meta_test.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bd1a0a9",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#chrom\tstart\tend\tpath\n",
      "chr1\t101384274\t104443097\tchr1/chr1_101384274_104443097.cor.xz,chr1/chr1_101384274_104443097.cor.xz.bim\n",
      "chr1\t104443097\t106225286\tchr1/chr1_104443097_106225286.cor.xz,chr1/chr1_104443097_106225286.cor.xz.bim\n",
      "chr1\t106225286\t109761915\tchr1/chr1_106225286_109761915.cor.xz,chr1/chr1_106225286_109761915.cor.xz.bim\n",
      "chr1\t109761915\t111483530\tchr1/chr1_109761915_111483530.cor.xz,chr1/chr1_109761915_111483530.cor.xz.bim\n",
      "chr1\t111483530\t113276642\tchr1/chr1_111483530_113276642.cor.xz,chr1/chr1_111483530_113276642.cor.xz.bim\n",
      "chr1\t113276642\t115338054\tchr1/chr1_113276642_115338054.cor.xz,chr1/chr1_113276642_115338054.cor.xz.bim\n",
      "chr1\t11328222\t12710318\tchr1/chr1_11328222_12710318.cor.xz,chr1/chr1_11328222_12710318.cor.xz.bim\n",
      "chr1\t115338054\t117562321\tchr1/chr1_115338054_117562321.cor.xz,chr1/chr1_115338054_117562321.cor.xz.bim\n",
      "chr1\t117562321\t119804207\tchr1/chr1_117562321_119804207.cor.xz,chr1/chr1_117562321_119804207.cor.xz.bim\n"
     ]
    }
   ],
   "source": [
    "# ld meta file\n",
    "head reference_data/ADSP_R4_EUR/ld_meta_file.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dcb4ce7",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\"#chr\" \"region_id\" \"TSS\" \"start\" \"end\" \"contexts\" \"original_data\"\n",
      "\"chr11\" \"ENSG00000073921\" 86069881 84957175 87360000 NA \"data/twas/ROSMAP_DeJager.chr11_ENSG00000073921.univariate_twas_weights.rds,data/twas/ROSMAP_DeJager.chr11_ENSG00000073921.multicontext_twas_weights.rds\"\n"
     ]
    }
   ],
   "source": [
    "# twas weight file\n",
    "head data/twas/mwe_twas_pipeline_test_small.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b60eec8c",
   "metadata": {},
   "source": [
    "To perform TWAS, you need to generate TWAS weight first.   \n",
    "see: \n",
    "- exercise notebook: `data/exercise_notebook/qtl_association_finemapping/2_finemapping.ipynb`\n",
    "- pipeline: `pipeline/mnm_regression.ipynb susie_twas`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30e457be",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 1\n",
      " $ ENSG00000002330:List of 10\n",
      "  ..$ AC_DeJager_eQTL_ENSG00000002330                :List of 7\n",
      "  .. ..$ susie_weights_intermediate:List of 4\n",
      "  .. .. ..$ mu                    : num [1:8, 1:7801] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:7801] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ lbf_variable          : num [1:8, 1:7801] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:7801] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ X_column_scale_factors: Named num [1:7801] 0.184 0.632 0.636 0.637 0.614 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:7801] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ pip                   : Named num [1:7801] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:7801] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ twas_weights              :List of 6\n",
      "  .. .. ..$ enet_weights   : num [1:7801, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:7801] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_weights  : num [1:7801, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:7801] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_weights: num [1:7801, 1] -1.69e-05 -1.04e-04 -1.65e-04 -1.48e-04 -9.45e-05 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:7801] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_weights: num [1:7801, 1] 8.28e-05 -4.19e-04 -4.27e-04 -6.72e-04 -5.37e-04 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:7801] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_weights  : num [1:7801, 1] -3.57e-06 -6.87e-05 -5.57e-05 -9.87e-05 -5.98e-05 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:7801] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_weights  : num [1:7801, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:7801] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_predictions          :List of 6\n",
      "  .. .. ..$ enet_predicted   : num [1:593, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:593] \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" \"SM-CJEG1\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_predicted  : num [1:593, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:593] \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" \"SM-CJEG1\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_predicted: num [1:593, 1] -0.008 0.1088 0.0192 -0.0433 0.0378 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:593] \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" \"SM-CJEG1\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_predicted: num [1:593, 1] -0.00862 0.24867 0.05078 -0.08654 0.07961 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:593] \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" \"SM-CJEG1\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_predicted  : num [1:593, 1] -0.00245 0.06505 0.01125 -0.0222 0.0232 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:593] \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" \"SM-CJEG1\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_predicted  : num [1:593, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:593] \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" \"SM-CJEG1\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_cv_result            :List of 4\n",
      "  .. .. ..$ sample_partition:'data.frame':\t593 obs. of  2 variables:\n",
      "  .. .. .. ..$ Sample: chr [1:593] \"SM-CJGGU\" \"SM-CJEKW\" \"SM-CJJ2O\" \"SM-CTDSB\" ...\n",
      "  .. .. .. ..$ Fold  : int [1:593] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. ..$ prediction      :List of 6\n",
      "  .. .. .. ..$ enet_predicted   : num [1:593, 1] 0.0155 0 0.0165 0.0123 0 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:593] \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" \"SM-CJEG1\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ lasso_predicted  : num [1:593, 1] 0.019 0 0.064 0 0 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:593] \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" \"SM-CJEG1\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_r_predicted: num [1:593, 1] -0.0149 0.1034 0.0707 -0.0632 0.0295 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:593] \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" \"SM-CJEG1\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_l_predicted: num [1:593, 1] -0.0337 0.2863 0.1065 -0.1708 0.0713 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:593] \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" \"SM-CJEG1\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ mrash_predicted  : num [1:593, 1] -0.00867 0.07222 0.02124 -0.03194 0.02293 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:593] \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" \"SM-CJEG1\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ susie_predicted  : num [1:593, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:593] \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" \"SM-CJEG1\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ performance     :List of 6\n",
      "  .. .. .. ..$ enet_performance   : num [1, 1:6] 0.046659 0.002177 0.000489 0.256613 0.428064 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ lasso_performance  : num [1, 1:6] 0.03256 0.00106 -0.00063 0.42875 0.43116 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_r_performance: num [1, 1:6] -0.028354 0.000804 -0.000887 0.490736 0.434856 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_l_performance: num [1, 1:6] -0.025919 0.000672 -0.001019 0.528729 0.444642 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ mrash_performance  : num [1, 1:6] -0.028474 0.000811 -0.00088 0.488892 0.429683 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ susie_performance  : logi [1, 1:6] NA NA NA NA NA NA\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. ..$ time_elapsed    : 'proc_time' Named num [1:5] 6.22 1.35 146.39 606.77 7.59\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ total_time_elapsed        : 'proc_time' Named num [1:5] 139.87 1.81 280.07 606.78 7.59\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ variant_names             : chr [1:7801] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ region_info               :List of 3\n",
      "  .. .. ..$ region_coord:'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 64284664\n",
      "  .. .. .. ..$ end  : int 64284665\n",
      "  .. .. ..$ grange      :'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 61200000\n",
      "  .. .. .. ..$ end  : int 65680000\n",
      "  .. .. ..$ region_name : chr [1:11] \"ENSG00000002330\" \"ENSG00000002330\" \"ENSG00000002330\" \"chr11_ENSG00000002330_Q92934\" ...\n",
      "  ..$ Ast_DeJager_eQTL_ENSG00000002330               :List of 7\n",
      "  .. ..$ susie_weights_intermediate:List of 4\n",
      "  .. .. ..$ mu                    : num [1:8, 1:8175] 1.10e-04 8.85e-05 7.16e-05 5.80e-05 4.70e-05 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:8175] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ lbf_variable          : num [1:8, 1:8175] -0.00653 -0.00527 -0.00426 -0.00345 -0.00279 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:8175] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ X_column_scale_factors: Named num [1:8175] 0.19 0.612 0.612 0.61 0.599 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:8175] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ pip                   : Named num [1:8175] 0.000975 0.000975 0.000975 0.000975 0.000975 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:8175] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ twas_weights              :List of 6\n",
      "  .. .. ..$ enet_weights   : num [1:8175, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8175] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_weights  : num [1:8175, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8175] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_weights: num [1:8175, 1] 6.07e-05 4.78e-05 8.08e-05 -2.13e-06 2.73e-05 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8175] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_weights: num [1:8175, 1] 6.45e-05 -1.46e-04 1.77e-04 -7.73e-05 1.88e-04 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8175] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_weights  : num [1:8175, 1] 2.17e-06 2.29e-06 5.94e-06 -4.86e-07 5.73e-06 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8175] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_weights  : num [1:8175, 1] 3.01e-07 6.70e-08 1.72e-07 -1.78e-08 1.67e-07 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8175] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_predictions          :List of 6\n",
      "  .. .. ..$ enet_predicted   : num [1:419, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_predicted  : num [1:419, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_predicted: num [1:419, 1] 0.0423 0.0498 -0.1254 -0.0142 -0.0325 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_predicted: num [1:419, 1] 0.1281 0.0995 -0.3543 -0.0696 -0.1212 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_predicted  : num [1:419, 1] 0.00433 0.00837 -0.02301 -0.00193 -0.00626 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_predicted  : num [1:419, 1] 1.58e-04 3.04e-04 -5.14e-04 -1.34e-04 -3.17e-05 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_cv_result            :List of 4\n",
      "  .. .. ..$ sample_partition:'data.frame':\t419 obs. of  2 variables:\n",
      "  .. .. .. ..$ Sample: chr [1:419] \"SM-CTDS5\" \"MAP50301099\" \"SM-CJFP7\" \"SM-CJGI4\" ...\n",
      "  .. .. .. ..$ Fold  : int [1:419] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. ..$ prediction      :List of 6\n",
      "  .. .. .. ..$ enet_predicted   : num [1:419, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ lasso_predicted  : num [1:419, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_r_predicted: num [1:419, 1] -0.0475 0.1175 -0.0243 -0.0566 0.0385 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_l_predicted: num [1:419, 1] -0.047 0.1793 -0.0256 -0.1788 0.0672 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ mrash_predicted  : num [1:419, 1] -0.0131 0.0309 -0.008 -0.0111 0.011 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ susie_predicted  : num [1:419, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ performance     :List of 6\n",
      "  .. .. .. ..$ enet_performance   : logi [1, 1:6] NA NA NA NA NA NA\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ lasso_performance  : logi [1, 1:6] NA NA NA NA NA NA\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_r_performance: num [1, 1:6] -0.08167 0.00667 0.00429 0.095 0.74496 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_l_performance: num [1, 1:6] -0.08287 0.00687 0.00449 0.09025 0.76052 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ mrash_performance  : num [1, 1:6] -0.06707 0.0045 0.00211 0.17058 0.73687 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ susie_performance  : num [1, 1:6] -0.041365 0.001711 -0.000683 0.398363 0.735802 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. ..$ time_elapsed    : 'proc_time' Named num [1:5] 4.43 1.04 130.44 512.83 9.01\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ total_time_elapsed        : 'proc_time' Named num [1:5] 112.14 1.31 237.65 512.83 9.02\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ variant_names             : chr [1:8175] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ region_info               :List of 3\n",
      "  .. .. ..$ region_coord:'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 64284664\n",
      "  .. .. .. ..$ end  : int 64284665\n",
      "  .. .. ..$ grange      :'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 61200000\n",
      "  .. .. .. ..$ end  : int 65680000\n",
      "  .. .. ..$ region_name : chr [1:11] \"ENSG00000002330\" \"ENSG00000002330\" \"ENSG00000002330\" \"chr11_ENSG00000002330_Q92934\" ...\n",
      "  ..$ DLPFC_Bennett_pQTL_chr11_ENSG00000002330_Q92934:List of 7\n",
      "  .. ..$ susie_weights_intermediate:List of 4\n",
      "  .. .. ..$ mu                    : num [1:8, 1:8136] 0.00988 0.00995 0.01005 0.01014 0.01022 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:8136] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ lbf_variable          : num [1:8, 1:8136] -0.0311 -0.0314 -0.0319 -0.0323 -0.0327 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:8136] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ X_column_scale_factors: Named num [1:8136] 0.141 0.648 0.651 0.64 0.626 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:8136] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ pip                   : Named num [1:8136] 0.000942 0.000807 0.000807 0.000807 0.000809 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:8136] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ twas_weights              :List of 6\n",
      "  .. .. ..$ enet_weights   : num [1:8136, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8136] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_weights  : num [1:8136, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8136] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_weights: num [1:8136, 1] 4.36e-05 -1.06e-05 -1.64e-05 -1.23e-04 4.89e-05 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8136] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_weights: num [1:8136, 1] 1.01e-04 7.81e-05 -3.62e-05 -5.74e-05 -9.74e-05 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8136] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_weights  : num [1:8136, 1] 1.44e-05 1.64e-06 -3.40e-07 5.73e-07 -5.22e-06 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8136] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_weights  : num [1:8136, 1] 6.78e-05 4.49e-07 -2.23e-07 2.15e-07 -1.95e-06 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8136] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_predictions          :List of 6\n",
      "  .. .. ..$ enet_predicted   : num [1:416, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:416] \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" \"MAP46251007\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_predicted  : num [1:416, 1] 0.0368 0.0184 0.0184 0.0184 0.0184 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:416] \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" \"MAP46251007\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_predicted: num [1:416, 1] -0.0534 0.0128 -0.113 -0.0417 -0.0772 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:416] \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" \"MAP46251007\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_predicted: num [1:416, 1] -0.11523 0.00147 -0.1901 -0.08675 -0.12961 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:416] \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" \"MAP46251007\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_predicted  : num [1:416, 1] -0.01725 0.00737 -0.04204 -0.01358 -0.02974 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:416] \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" \"MAP46251007\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_predicted  : num [1:416, 1] -0.0085 -0.00202 -0.02284 -0.00931 -0.01425 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:416] \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" \"MAP46251007\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_cv_result            :List of 4\n",
      "  .. .. ..$ sample_partition:'data.frame':\t416 obs. of  2 variables:\n",
      "  .. .. .. ..$ Sample: chr [1:416] \"SM-CJIZZ\" \"SM-CJGID\" \"SM-CJK5C\" \"SM-CTED8\" ...\n",
      "  .. .. .. ..$ Fold  : int [1:416] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. ..$ prediction      :List of 6\n",
      "  .. .. .. ..$ enet_predicted   : num [1:416, 1] 0.219 0.13 0 0.144 0 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:416] \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" \"MAP46251007\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ lasso_predicted  : num [1:416, 1] 0.1417 0.0765 0 0.0778 0.0131 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:416] \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" \"MAP46251007\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_r_predicted: num [1:416, 1] -0.0305 -0.0755 -0.0939 -0.0925 -0.0606 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:416] \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" \"MAP46251007\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_l_predicted: num [1:416, 1] -0.0641 -0.0953 -0.1739 -0.1285 -0.0954 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:416] \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" \"MAP46251007\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ mrash_predicted  : num [1:416, 1] 0.00199 -0.01427 -0.01583 -0.01669 -0.01244 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:416] \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" \"MAP46251007\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ susie_predicted  : num [1:416, 1] -0.00065 -0.005202 -0.000703 -0.004044 0 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:416] \"MAP26637867\" \"MAP34726040\" \"MAP46246604\" \"MAP46251007\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ performance     :List of 6\n",
      "  .. .. .. ..$ enet_performance   : num [1, 1:6] 0.03522 0.00124 -0.00117 0.47374 0.66505 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ lasso_performance  : num [1, 1:6] 0.02611 0.000682 -0.001732 0.595398 0.664557 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_r_performance: num [1, 1:6] 0.040788 0.001664 -0.000748 0.406678 0.666476 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_l_performance: num [1, 1:6] 0.03549 0.00126 -0.00115 0.4704 0.67449 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ mrash_performance  : num [1, 1:6] 0.022508 0.000507 -0.001908 0.647132 0.663256 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ susie_performance  : num [1, 1:6] 2.70e-04 7.27e-08 -2.42e-03 9.96e-01 6.63e-01 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. ..$ time_elapsed    : 'proc_time' Named num [1:5] 5 1.14 131.88 545.53 10.26\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ total_time_elapsed        : 'proc_time' Named num [1:5] 102.42 1.23 228.77 545.53 10.26\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ variant_names             : chr [1:8136] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ region_info               :List of 3\n",
      "  .. .. ..$ region_coord:'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 64284664\n",
      "  .. .. .. ..$ end  : int 64284665\n",
      "  .. .. ..$ grange      :'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 61200000\n",
      "  .. .. .. ..$ end  : int 65680000\n",
      "  .. .. ..$ region_name : chr [1:11] \"ENSG00000002330\" \"ENSG00000002330\" \"ENSG00000002330\" \"chr11_ENSG00000002330_Q92934\" ...\n",
      "  ..$ DLPFC_DeJager_eQTL_ENSG00000002330             :List of 7\n",
      "  .. ..$ susie_weights_intermediate:List of 4\n",
      "  .. .. ..$ mu                    : num [1:8, 1:7249] -2.56e-07 -2.41e-07 -2.24e-07 -2.06e-07 -1.88e-07 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:7249] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202736:C:A\" ...\n",
      "  .. .. ..$ lbf_variable          : num [1:8, 1:7249] -0.00651 -0.00613 -0.00571 -0.00526 -0.00481 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:7249] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202736:C:A\" ...\n",
      "  .. .. ..$ X_column_scale_factors: Named num [1:7249] 0.176 0.645 0.647 0.628 0.63 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:7249] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202736:C:A\" ...\n",
      "  .. .. ..$ pip                   : Named num [1:7249] 0.0011 0.0011 0.0011 0.00111 0.00111 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:7249] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202736:C:A\" ...\n",
      "  .. ..$ twas_weights              :List of 6\n",
      "  .. .. ..$ enet_weights   : num [1:7249, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:7249] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202736:C:A\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_weights  : num [1:7249, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:7249] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202736:C:A\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_weights: num [1:7249, 1] -8.18e-05 -1.30e-04 -7.84e-05 -1.53e-04 -2.03e-04 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:7249] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202736:C:A\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_weights: num [1:7249, 1] 9.65e-05 -6.97e-04 -3.23e-04 -7.64e-04 -8.71e-04 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:7249] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202736:C:A\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_weights  : num [1:7249, 1] 2.73e-07 -1.35e-04 -8.96e-05 -1.86e-04 -2.04e-04 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:7249] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202736:C:A\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_weights  : num [1:7249, 1] -1.23e-09 -1.21e-07 -7.76e-08 -1.70e-07 -1.87e-07 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:7249] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202736:C:A\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_predictions          :List of 6\n",
      "  .. .. ..$ enet_predicted   : num [1:784, 1] -0.00293 -0.00293 -0.00586 0 -0.00293 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:784] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_predicted  : num [1:784, 1] -0.00794 -0.00794 -0.01589 0 -0.00794 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:784] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_predicted: num [1:784, 1] -0.0118 -0.011 -0.0401 -0.0668 0.0384 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:784] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_predicted: num [1:784, 1] -0.01 -0.0363 -0.0821 -0.1522 0.0856 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:784] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_predicted  : num [1:784, 1] -0.00151 -0.00654 -0.02212 -0.03848 0.02412 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:784] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_predicted  : num [1:784, 1] -4.92e-05 -5.39e-05 -6.18e-05 -1.66e-04 2.64e-05 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:784] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_cv_result            :List of 4\n",
      "  .. .. ..$ sample_partition:'data.frame':\t784 obs. of  2 variables:\n",
      "  .. .. .. ..$ Sample: chr [1:784] \"SM-CJFNI\" \"SM-CTEG8\" \"SM-CJFM3\" \"SM-CTECR\" ...\n",
      "  .. .. .. ..$ Fold  : int [1:784] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. ..$ prediction      :List of 6\n",
      "  .. .. .. ..$ enet_predicted   : num [1:784, 1] -0.02928 0 -0.01656 0.00267 -0.00912 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:784] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ lasso_predicted  : num [1:784, 1] -0.03643 0 -0.00706 0 -0.0059 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:784] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_r_predicted: num [1:784, 1] -0.0561 0.0014 -0.0393 -0.0627 0.0846 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:784] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_l_predicted: num [1:784, 1] -0.0384 0.0472 -0.109 -0.1753 0.1145 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:784] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ mrash_predicted  : num [1:784, 1] -0.00853 0.01095 -0.03135 -0.04856 0.03705 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:784] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ susie_predicted  : num [1:784, 1] -0.00421 0 0 0 0 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:784] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFS\" \"SM-CJEFU\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ performance     :List of 6\n",
      "  .. .. .. ..$ enet_performance   : num [1, 1:6] 0.026694 0.000713 -0.000565 0.455442 0.216162 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ lasso_performance  : num [1, 1:6] 0.01748 0.000306 -0.000973 0.625047 0.2164 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_r_performance: num [1, 1:6] -0.041746 0.001743 0.000466 0.242992 0.226976 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_l_performance: num [1, 1:6] -0.032785 0.001075 -0.000203 0.359261 0.241253 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ mrash_performance  : num [1, 1:6] -3.59e-02 1.29e-03 1.24e-05 3.15e-01 2.19e-01 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ susie_performance  : num [1, 1:6] -0.06631 0.0044 0.00312 0.06349 0.21619 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. ..$ time_elapsed    : 'proc_time' Named num [1:5] 6 1.48 160.98 679.54 8.11\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ total_time_elapsed        : 'proc_time' Named num [1:5] 152.25 1.7 306.95 679.54 8.12\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ variant_names             : chr [1:7249] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202736:C:A\" ...\n",
      "  .. ..$ region_info               :List of 3\n",
      "  .. .. ..$ region_coord:'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 64284664\n",
      "  .. .. .. ..$ end  : int 64284665\n",
      "  .. .. ..$ grange      :'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 61200000\n",
      "  .. .. .. ..$ end  : int 65680000\n",
      "  .. .. ..$ region_name : chr [1:11] \"ENSG00000002330\" \"ENSG00000002330\" \"ENSG00000002330\" \"chr11_ENSG00000002330_Q92934\" ...\n",
      "  ..$ Exc_DeJager_eQTL_ENSG00000002330               :List of 7\n",
      "  .. ..$ susie_weights_intermediate:List of 5\n",
      "  .. .. ..$ mu                    : num [1:8, 1:8186] -0.00149 -0.0015 -0.00152 -0.00153 -0.00154 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ lbf_variable          : num [1:8, 1:8186] -0.424 -0.433 -0.444 -0.452 -0.454 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ X_column_scale_factors: Named num [1:8186] 0.191 0.626 0.63 0.624 0.609 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ pip                   : Named num [1:8186] 0.000597 0.000678 0.000652 0.000685 0.000663 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ top_loci              :'data.frame':\t3 obs. of  8 variables:\n",
      "  .. .. .. ..$ variant_id      : chr [1:3] \"chr11:63191856:C:T\" \"chr11:63375056:T:C\" \"chr11:63377192:A:G\"\n",
      "  .. .. .. ..$ betahat         : num [1:3] 0.183 0.173 0.191\n",
      "  .. .. .. ..$ sebetahat       : num [1:3] 0.0447 0.0421 0.0453\n",
      "  .. .. .. ..$ maf             : num [1:3] 0.0644 0.0721 0.0586\n",
      "  .. .. .. ..$ pip             : num [1:3] 0.0273 0.0301 0.0343\n",
      "  .. .. .. ..$ cs_coverage_0.95: num [1:3] 0 0 0\n",
      "  .. .. .. ..$ cs_coverage_0.7 : num [1:3] 0 0 0\n",
      "  .. .. .. ..$ cs_coverage_0.5 : num [1:3] 0 0 0\n",
      "  .. ..$ twas_weights              :List of 6\n",
      "  .. .. ..$ enet_weights   : num [1:8186, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_weights  : num [1:8186, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_weights: num [1:8186, 1] -8.25e-05 1.40e-04 1.30e-04 1.62e-04 9.88e-05 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_weights: num [1:8186, 1] 0.000017 0.00061 0.000282 0.000368 0.000116 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_weights  : num [1:8186, 1] -1.03e-05 8.35e-05 6.19e-05 8.87e-05 7.32e-05 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_weights  : num [1:8186, 1] -4.75e-06 6.37e-06 5.10e-06 6.69e-06 5.83e-06 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_predictions          :List of 6\n",
      "  .. .. ..$ enet_predicted   : num [1:419, 1] 0.00282 -0.018 -0.04108 -0.01217 0.02921 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_predicted  : num [1:419, 1] -0.00361 -0.02189 -0.05139 -0.0186 0.0238 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_predicted: num [1:419, 1] -0.0159 -0.1136 -0.2165 -0.0327 -0.0864 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_predicted: num [1:419, 1] -0.0229 -0.16912 -0.27029 0.00632 -0.14738 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_predicted  : num [1:419, 1] -0.0142 -0.0512 -0.0809 -0.0108 -0.0452 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_predicted  : num [1:419, 1] 0.00361 -0.01634 -0.03554 -0.00673 -0.0087 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_cv_result            :List of 4\n",
      "  .. .. ..$ sample_partition:'data.frame':\t419 obs. of  2 variables:\n",
      "  .. .. .. ..$ Sample: chr [1:419] \"SM-CJIY2\" \"SM-CJJ2E\" \"SM-CJIXK\" \"SM-CJJ1F\" ...\n",
      "  .. .. .. ..$ Fold  : int [1:419] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. ..$ prediction      :List of 6\n",
      "  .. .. .. ..$ enet_predicted   : num [1:419, 1] -0.0815 0 -0.1203 -0.0832 0 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ lasso_predicted  : num [1:419, 1] -0.102 0 -0.147 -0.108 0 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_r_predicted: num [1:419, 1] -0.1504 -0.0929 -0.391 -0.1198 -0.1189 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_l_predicted: num [1:419, 1] -0.158 -0.131 -0.427 -0.113 -0.165 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ mrash_predicted  : num [1:419, 1] -0.0571 -0.0418 -0.1499 -0.0427 -0.0499 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ susie_predicted  : num [1:419, 1] -0.01372 -0.0014 -0.03181 -0.01484 0.00102 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ performance     :List of 6\n",
      "  .. .. .. ..$ enet_performance   : num [1, 1:6] 0.07493 0.00562 0.00323 0.12566 0.3023 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ lasso_performance  : num [1, 1:6] 0.06284 0.00395 0.00156 0.19924 0.30339 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_r_performance: num [1, 1:6] 0.10674 0.01139 0.00902 0.02891 0.3486 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_l_performance: num [1, 1:6] 0.1185 0.014 0.0117 0.0152 0.3567 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ mrash_performance  : num [1, 1:6] 0.1354 0.0183 0.016 0.0055 0.3055 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ susie_performance  : num [1, 1:6] 0.07739 0.00599 0.0036 0.11372 0.30213 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. ..$ time_elapsed    : 'proc_time' Named num [1:5] 4.88 1.16 128.3 549.91 7.94\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ total_time_elapsed        : 'proc_time' Named num [1:5] 100.1 1.32 222.79 549.91 7.94\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ variant_names             : chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ region_info               :List of 3\n",
      "  .. .. ..$ region_coord:'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 64284664\n",
      "  .. .. .. ..$ end  : int 64284665\n",
      "  .. .. ..$ grange      :'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 61200000\n",
      "  .. .. .. ..$ end  : int 65680000\n",
      "  .. .. ..$ region_name : chr [1:11] \"ENSG00000002330\" \"ENSG00000002330\" \"ENSG00000002330\" \"chr11_ENSG00000002330_Q92934\" ...\n",
      "  ..$ Inh_DeJager_eQTL_ENSG00000002330               :List of 7\n",
      "  .. ..$ susie_weights_intermediate:List of 4\n",
      "  .. .. ..$ mu                    : num [1:8, 1:8186] -0.00603 -0.0061 -0.00618 -0.00625 -0.00631 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ lbf_variable          : num [1:8, 1:8186] -0.102 -0.104 -0.105 -0.107 -0.108 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ X_column_scale_factors: Named num [1:8186] 0.188 0.61 0.616 0.609 0.599 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ pip                   : Named num [1:8186] 0.000873 0.000883 0.000833 0.001084 0.000969 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ twas_weights              :List of 6\n",
      "  .. .. ..$ enet_weights   : num [1:8186, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_weights  : num [1:8186, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_weights: num [1:8186, 1] -0.000189 0.000136 0.000075 0.000186 0.000144 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_weights: num [1:8186, 1] -4.92e-05 2.37e-04 5.72e-05 6.88e-04 3.02e-04 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_weights  : num [1:8186, 1] -1.05e-05 2.00e-05 1.03e-05 4.65e-05 3.24e-05 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_weights  : num [1:8186, 1] -2.89e-05 9.75e-06 5.08e-06 2.21e-05 1.57e-05 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_predictions          :List of 6\n",
      "  .. .. ..$ enet_predicted   : num [1:419, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_predicted  : num [1:419, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_predicted: num [1:419, 1] 0.0152 -0.136 -0.0317 -0.0358 0.0186 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_predicted: num [1:419, 1] 0.0113 -0.2653 -0.0876 -0.0738 0.0207 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_predicted  : num [1:419, 1] 0.000517 -0.037556 -0.012399 -0.012414 -0.003268 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_predicted  : num [1:419, 1] 0.00121 -0.02237 -0.00438 -0.00583 0.00137 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_cv_result            :List of 4\n",
      "  .. .. ..$ sample_partition:'data.frame':\t419 obs. of  2 variables:\n",
      "  .. .. .. ..$ Sample: chr [1:419] \"SM-CJJ1R\" \"SM-CJGML\" \"SM-CTDQK\" \"SM-CTECU\" ...\n",
      "  .. .. .. ..$ Fold  : int [1:419] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. ..$ prediction      :List of 6\n",
      "  .. .. .. ..$ enet_predicted   : num [1:419, 1] 0 0 0 0 0 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ lasso_predicted  : num [1:419, 1] 0 0 0 0 -0.0424 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_r_predicted: num [1:419, 1] -0.00837 -0.06691 -0.03588 -0.05745 -0.02289 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_l_predicted: num [1:419, 1] -0.0202 -0.1365 -0.043 -0.1156 -0.0275 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ mrash_predicted  : num [1:419, 1] -0.00241 -0.01475 -0.01804 -0.0103 -0.00906 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ susie_predicted  : num [1:419, 1] 0 0 -0.00158 0 0 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ performance     :List of 6\n",
      "  .. .. .. ..$ enet_performance   : num [1, 1:6] -0.08017 0.00643 0.00404 0.10127 0.61137 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ lasso_performance  : num [1, 1:6] -0.07706 0.00594 0.00355 0.11526 0.6127 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_r_performance: num [1, 1:6] -0.023057 0.000532 -0.001865 0.637919 0.619349 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_l_performance: num [1, 1:6] -3.80e-03 1.45e-05 -2.38e-03 9.38e-01 6.24e-01 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ mrash_performance  : num [1, 1:6] -0.08674 0.00752 0.00514 0.07614 0.62404 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ susie_performance  : num [1, 1:6] -0.07814 0.00611 0.00372 0.11025 0.62242 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. ..$ time_elapsed    : 'proc_time' Named num [1:5] 4.27 1.07 130.84 511.96 8.52\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ total_time_elapsed        : 'proc_time' Named num [1:5] 99.34 1.2 225.23 511.96 8.52\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ variant_names             : chr [1:8186] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ region_info               :List of 3\n",
      "  .. .. ..$ region_coord:'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 64284664\n",
      "  .. .. .. ..$ end  : int 64284665\n",
      "  .. .. ..$ grange      :'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 61200000\n",
      "  .. .. .. ..$ end  : int 65680000\n",
      "  .. .. ..$ region_name : chr [1:11] \"ENSG00000002330\" \"ENSG00000002330\" \"ENSG00000002330\" \"chr11_ENSG00000002330_Q92934\" ...\n",
      "  ..$ OPC_DeJager_eQTL_ENSG00000002330               :List of 7\n",
      "  .. ..$ susie_weights_intermediate:List of 5\n",
      "  .. .. ..$ mu                    : num [1:8, 1:8181] -0.03 -0.0218 -0.0164 -0.0162 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:8181] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ lbf_variable          : num [1:8, 1:8181] -1.058 -1.018 -0.198 -0.193 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:8181] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ X_column_scale_factors: Named num [1:8181] 0.191 0.62 0.624 0.621 0.603 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:8181] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ pip                   : Named num [1:8181] 0.000228 0.000194 0.000192 0.000195 0.000192 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:8181] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ top_loci              :'data.frame':\t67 obs. of  8 variables:\n",
      "  .. .. .. ..$ variant_id      : chr [1:67] \"chr11:62108885:A:G\" \"chr11:62118627:A:C\" \"chr11:62118673:C:G\" \"chr11:62119034:T:C\" ...\n",
      "  .. .. .. ..$ betahat         : num [1:67] 0.253 0.258 0.258 0.258 -0.3 ...\n",
      "  .. .. .. ..$ sebetahat       : num [1:67] 0.0578 0.058 0.058 0.058 0.0694 ...\n",
      "  .. .. .. ..$ maf             : num [1:67] 0.467 0.468 0.468 0.468 0.255 ...\n",
      "  .. .. .. ..$ pip             : num [1:67] 0.0272 0.037 0.037 0.037 0.0178 ...\n",
      "  .. .. .. ..$ cs_coverage_0.95: num [1:67] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..$ cs_coverage_0.7 : num [1:67] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. .. ..$ cs_coverage_0.5 : num [1:67] 1 1 1 1 0 1 1 1 1 1 ...\n",
      "  .. ..$ twas_weights              :List of 6\n",
      "  .. .. ..$ enet_weights   : num [1:8181, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8181] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_weights  : num [1:8181, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8181] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_weights: num [1:8181, 1] -1.66e-04 -6.73e-05 -3.93e-05 -4.48e-05 4.17e-05 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8181] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_weights: num [1:8181, 1] -0.000176 -0.000147 0.000212 -0.000193 -0.000173 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8181] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_weights  : num [1:8181, 1] -3.13e-05 -9.58e-06 2.19e-06 -9.13e-06 6.42e-06 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8181] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_weights  : num [1:8181, 1] -2.05e-05 -1.35e-06 6.04e-08 -1.59e-06 3.03e-07 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8181] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_predictions          :List of 6\n",
      "  .. .. ..$ enet_predicted   : num [1:418, 1] -0.16393 0.00896 -0.01117 0.05934 -0.0455 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:418] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_predicted  : num [1:418, 1] -0.133 0.00294 0.04294 0.06893 -0.02126 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:418] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_predicted: num [1:418, 1] -0.167416 0.000255 -0.023756 0.157267 -0.131488 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:418] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_predicted: num [1:418, 1] -0.14552 -0.00971 -0.04828 0.11819 -0.1382 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:418] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_predicted  : num [1:418, 1] -0.07076 0.00649 0.01253 0.07179 -0.04131 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:418] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_predicted  : num [1:418, 1] -0.0325 0.1619 0.1439 0.322 0.1233 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:418] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_cv_result            :List of 4\n",
      "  .. .. ..$ sample_partition:'data.frame':\t418 obs. of  2 variables:\n",
      "  .. .. .. ..$ Sample: chr [1:418] \"SM-CTEFX\" \"SM-CJJ1C\" \"SM-CJIXY\" \"SM-CJIZK\" ...\n",
      "  .. .. .. ..$ Fold  : int [1:418] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. ..$ prediction      :List of 6\n",
      "  .. .. .. ..$ enet_predicted   : num [1:418, 1] -0.1445 -0.1952 0.0708 -0.2215 0.021 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:418] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ lasso_predicted  : num [1:418, 1] -0.2271 -0.197 0.0241 -0.2223 0.0241 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:418] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_r_predicted: num [1:418, 1] -0.1746 -0.0388 0.0766 0.0487 -0.1264 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:418] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_l_predicted: num [1:418, 1] -0.2036 -0.0853 -0.0149 0.0463 -0.2129 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:418] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ mrash_predicted  : num [1:418, 1] -0.06598 -0.04164 0.03112 0.00339 -0.06175 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:418] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ susie_predicted  : num [1:418, 1] -0.04246 -0.02805 0.14454 0.00306 0.12449 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:418] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ performance     :List of 6\n",
      "  .. .. .. ..$ enet_performance   : num [1, 1:6] 0.1124 0.0126 0.0103 0.0215 0.8323 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ lasso_performance  : num [1, 1:6] 0.1128 0.0127 0.0104 0.0211 0.8347 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_r_performance: num [1, 1:6] 0.14114 0.01992 0.01756 0.00384 0.82546 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_l_performance: num [1, 1:6] 0.13555 0.01837 0.01601 0.00551 0.82805 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ mrash_performance  : num [1, 1:6] 0.15439 0.02384 0.02149 0.00154 0.82514 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ susie_performance  : num [1, 1:6] 0.1236 0.0153 0.0129 0.0115 0.8279 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. ..$ time_elapsed    : 'proc_time' Named num [1:5] 4.72 1.2 141.75 555.89 10.57\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ total_time_elapsed        : 'proc_time' Named num [1:5] 96.25 1.26 232.43 555.89 10.57\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ variant_names             : chr [1:8181] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ region_info               :List of 3\n",
      "  .. .. ..$ region_coord:'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 64284664\n",
      "  .. .. .. ..$ end  : int 64284665\n",
      "  .. .. ..$ grange      :'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 61200000\n",
      "  .. .. .. ..$ end  : int 65680000\n",
      "  .. .. ..$ region_name : chr [1:11] \"ENSG00000002330\" \"ENSG00000002330\" \"ENSG00000002330\" \"chr11_ENSG00000002330_Q92934\" ...\n",
      "  ..$ Oli_DeJager_eQTL_ENSG00000002330               :List of 7\n",
      "  .. ..$ susie_weights_intermediate:List of 4\n",
      "  .. .. ..$ mu                    : num [1:8, 1:8170] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:8170] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ lbf_variable          : num [1:8, 1:8170] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:8170] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ X_column_scale_factors: Named num [1:8170] 0.186 0.61 0.615 0.607 0.595 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:8170] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ pip                   : Named num [1:8170] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:8170] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ twas_weights              :List of 6\n",
      "  .. .. ..$ enet_weights   : num [1:8170, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8170] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_weights  : num [1:8170, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8170] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_weights: num [1:8170, 1] -2.41e-05 1.07e-04 1.77e-04 1.47e-04 7.90e-06 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8170] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_weights: num [1:8170, 1] 6.12e-05 1.28e-04 3.29e-04 1.40e-04 -1.88e-05 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8170] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_weights  : num [1:8170, 1] 9.24e-07 3.31e-06 6.03e-06 4.03e-06 1.65e-06 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8170] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_weights  : num [1:8170, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8170] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_predictions          :List of 6\n",
      "  .. .. ..$ enet_predicted   : num [1:419, 1] -0.0667 -0.0634 -0.1099 -0.051 -0.0344 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_predicted  : num [1:419, 1] -0.0978 -0.0836 -0.1487 -0.0741 -0.0419 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_predicted: num [1:419, 1] -0.04432 0.00628 -0.00523 0.038 0.04927 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_predicted: num [1:419, 1] -0.039148 0.000741 0.027977 0.055285 0.087737 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_predicted  : num [1:419, 1] -0.004074 -0.000737 -0.002653 0.003627 0.002466 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_predicted  : num [1:419, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_cv_result            :List of 4\n",
      "  .. .. ..$ sample_partition:'data.frame':\t419 obs. of  2 variables:\n",
      "  .. .. .. ..$ Sample: chr [1:419] \"SM-CJEGB\" \"SM-CJIXC\" \"MAP89164957\" \"SM-CTEMY\" ...\n",
      "  .. .. .. ..$ Fold  : int [1:419] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. ..$ prediction      :List of 6\n",
      "  .. .. .. ..$ enet_predicted   : num [1:419, 1] 0.0121 -0.0233 0 0 0.0241 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ lasso_predicted  : num [1:419, 1] 0 0 0 0 0 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_r_predicted: num [1:419, 1] -0.07591 -0.01895 0.00988 0.00198 0.05214 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_l_predicted: num [1:419, 1] -0.1066 -0.0345 0.0312 -0.0249 0.1111 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ mrash_predicted  : num [1:419, 1] -5.79e-03 -7.28e-03 1.11e-03 1.74e-05 3.25e-03 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ susie_predicted  : num [1:419, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:419] \"MAP15387421\" \"MAP46246604\" \"MAP50104134\" \"MAP50104846\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ performance     :List of 6\n",
      "  .. .. .. ..$ enet_performance   : num [1, 1:6] 1.06e-03 1.13e-06 -2.40e-03 9.83e-01 7.09e-01 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ lasso_performance  : num [1, 1:6] -0.05998 0.0036 0.00121 0.22052 0.71201 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_r_performance: num [1, 1:6] -0.08517 0.00725 0.00487 0.08162 0.71471 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_l_performance: num [1, 1:6] -0.08997 0.00809 0.00572 0.06579 0.72448 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ mrash_performance  : num [1, 1:6] -0.10093 0.01019 0.00781 0.03891 0.70949 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ susie_performance  : num [1, 1:6] -0.14428 0.02082 0.01847 0.00308 0.70902 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. ..$ time_elapsed    : 'proc_time' Named num [1:5] 4.2 1.07 123.46 471.24 8.03\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ total_time_elapsed        : 'proc_time' Named num [1:5] 93.86 1.21 212.3 471.24 8.03\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ variant_names             : chr [1:8170] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ region_info               :List of 3\n",
      "  .. .. ..$ region_coord:'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 64284664\n",
      "  .. .. .. ..$ end  : int 64284665\n",
      "  .. .. ..$ grange      :'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 61200000\n",
      "  .. .. .. ..$ end  : int 65680000\n",
      "  .. .. ..$ region_name : chr [1:11] \"ENSG00000002330\" \"ENSG00000002330\" \"ENSG00000002330\" \"chr11_ENSG00000002330_Q92934\" ...\n",
      "  ..$ PCC_DeJager_eQTL_ENSG00000002330               :List of 7\n",
      "  .. ..$ susie_weights_intermediate:List of 5\n",
      "  .. .. ..$ mu                    : num [1:8, 1:8245] 0.0164 0.0172 0.017 0.0166 0.0162 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:8245] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ lbf_variable          : num [1:8, 1:8245] 0.111 0.102 0.104 0.109 0.113 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:8245] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ X_column_scale_factors: Named num [1:8245] 0.16 0.636 0.639 0.648 0.618 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:8245] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ pip                   : Named num [1:8245] 0.001062 0.000638 0.000639 0.000619 0.000652 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:8245] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ top_loci              :'data.frame':\t1 obs. of  8 variables:\n",
      "  .. .. .. ..$ variant_id      : chr \"chr11:64851296:T:C\"\n",
      "  .. .. .. ..$ betahat         : num -0.146\n",
      "  .. .. .. ..$ sebetahat       : num 0.0338\n",
      "  .. .. .. ..$ maf             : num 0.291\n",
      "  .. .. .. ..$ pip             : num 0.0604\n",
      "  .. .. .. ..$ cs_coverage_0.95: num 0\n",
      "  .. .. .. ..$ cs_coverage_0.7 : num 0\n",
      "  .. .. .. ..$ cs_coverage_0.5 : num 0\n",
      "  .. ..$ twas_weights              :List of 6\n",
      "  .. .. ..$ enet_weights   : num [1:8245, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8245] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_weights  : num [1:8245, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8245] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_weights: num [1:8245, 1] -4.02e-05 3.26e-05 -1.19e-04 -1.87e-04 -2.34e-06 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8245] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_weights: num [1:8245, 1] 2.26e-04 9.51e-05 -2.62e-04 -2.23e-04 1.19e-05 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8245] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_weights  : num [1:8245, 1] 3.27e-05 1.64e-05 -2.36e-05 -1.02e-05 2.16e-05 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8245] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_weights  : num [1:8245, 1] 1.07e-04 4.11e-06 -4.23e-06 -1.62e-06 5.52e-06 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8245] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_predictions          :List of 6\n",
      "  .. .. ..$ enet_predicted   : num [1:441, 1] -0.0404 -0.0525 -0.0568 0.0718 0.0175 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:441] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFU\" \"SM-CJEFX\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_predicted  : num [1:441, 1] -0.08318 -0.09599 -0.10165 0.09467 0.00625 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:441] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFU\" \"SM-CJEFX\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_predicted: num [1:441, 1] 0.1865 0.0296 -0.0546 0.4432 0.1951 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:441] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFU\" \"SM-CJEFX\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_predicted: num [1:441, 1] 0.0844 -0.0438 -0.1552 0.3104 0.0785 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:441] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFU\" \"SM-CJEFX\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_predicted  : num [1:441, 1] 0.0462 -0.00143 -0.02616 0.09915 0.03712 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:441] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFU\" \"SM-CJEFX\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_predicted  : num [1:441, 1] 0.05027 0.01812 0.00361 0.10016 0.05817 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:441] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFU\" \"SM-CJEFX\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_cv_result            :List of 4\n",
      "  .. .. ..$ sample_partition:'data.frame':\t441 obs. of  2 variables:\n",
      "  .. .. .. ..$ Sample: chr [1:441] \"SM-CJGHM\" \"SM-CTEN5\" \"SM-CTEMW\" \"SM-CJIYD\" ...\n",
      "  .. .. .. ..$ Fold  : int [1:441] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. ..$ prediction      :List of 6\n",
      "  .. .. .. ..$ enet_predicted   : num [1:441, 1] -0.0668 -0.0557 -0.0613 -0.0052 -0.0174 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:441] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFU\" \"SM-CJEFX\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ lasso_predicted  : num [1:441, 1] -0.1074 -0.0957 0 0 -0.0226 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:441] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFU\" \"SM-CJEFX\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_r_predicted: num [1:441, 1] 0.1536 0.0389 -0.0706 0.0975 0.0591 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:441] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFU\" \"SM-CJEFX\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_l_predicted: num [1:441, 1] 0.1642 0.0281 -0.1263 0.1564 0.023 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:441] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFU\" \"SM-CJEFX\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ mrash_predicted  : num [1:441, 1] 0.0333 0.0101 -0.0468 0.0581 0.0156 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:441] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFU\" \"SM-CJEFX\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ susie_predicted  : num [1:441, 1] 0.00894 0.00467 0 0 0 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:441] \"SM-CJEFQ\" \"SM-CJEFR\" \"SM-CJEFU\" \"SM-CJEFX\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ performance     :List of 6\n",
      "  .. .. .. ..$ enet_performance   : num [1, 1:6] 0.08842 0.00782 0.00556 0.06357 0.43189 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ lasso_performance  : num [1, 1:6] 0.06116 0.00374 0.00147 0.19991 0.43385 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_r_performance: num [1, 1:6] -0.01037 0.000108 -0.00217 0.828078 0.465019 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_l_performance: num [1, 1:6] 1.07e-03 1.15e-06 -2.28e-03 9.82e-01 4.66e-01 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ mrash_performance  : num [1, 1:6] 0.015419 0.000238 -0.00204 0.746767 0.433475 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ susie_performance  : num [1, 1:6] -0.052478 0.002754 0.000482 0.271473 0.436221 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. ..$ time_elapsed    : 'proc_time' Named num [1:5] 4.85 1.17 134.33 592.55 9.48\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ total_time_elapsed        : 'proc_time' Named num [1:5] 99.89 1.22 228.62 592.56 9.48\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ variant_names             : chr [1:8245] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ region_info               :List of 3\n",
      "  .. .. ..$ region_coord:'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 64284664\n",
      "  .. .. .. ..$ end  : int 64284665\n",
      "  .. .. ..$ grange      :'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 61200000\n",
      "  .. .. .. ..$ end  : int 65680000\n",
      "  .. .. ..$ region_name : chr [1:11] \"ENSG00000002330\" \"ENSG00000002330\" \"ENSG00000002330\" \"chr11_ENSG00000002330_Q92934\" ...\n",
      "  ..$ monocyte_ROSMAP_eQTL_ENSG00000002330           :List of 7\n",
      "  .. ..$ susie_weights_intermediate:List of 7\n",
      "  .. .. ..$ mu                    : num [1:8, 1:8057] 0.026814 0.001895 0.000646 0.000252 0.000101 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:8057] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ lbf_variable          : num [1:8, 1:8057] -0.18633 0.07556 0.02597 0.01014 0.00409 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr [1:8057] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ X_column_scale_factors: Named num [1:8057] 0.15 0.589 0.59 0.576 0.59 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:8057] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ pip                   : Named num [1:8057] 0.000886 0.00087 0.000873 0.000863 0.000866 ...\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:8057] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. ..$ cs_variants           :List of 1\n",
      "  .. .. .. ..$ L1: chr [1:24] \"chr11:64110878:C:G\" \"chr11:64112182:C:A\" \"chr11:64118297:C:T\" \"chr11:64122277:T:G\" ...\n",
      "  .. .. ..$ cs_purity             :'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ min.abs.corr   : num 0.921\n",
      "  .. .. .. ..$ mean.abs.corr  : num 0.987\n",
      "  .. .. .. ..$ median.abs.corr: num 1\n",
      "  .. .. ..$ top_loci              :'data.frame':\t24 obs. of  8 variables:\n",
      "  .. .. .. ..$ variant_id      : chr [1:24] \"chr11:64110878:C:G\" \"chr11:64112182:C:A\" \"chr11:64118297:C:T\" \"chr11:64122277:T:G\" ...\n",
      "  .. .. .. ..$ betahat         : num [1:24] 0.25 0.25 0.25 0.25 0.25 ...\n",
      "  .. .. .. ..$ sebetahat       : num [1:24] 0.0514 0.0514 0.0514 0.0514 0.0514 ...\n",
      "  .. .. .. ..$ maf             : num [1:24] 0.0708 0.0708 0.0708 0.0708 0.0708 ...\n",
      "  .. .. .. ..$ pip             : num [1:24] 0.0463 0.0463 0.0463 0.0463 0.0463 ...\n",
      "  .. .. .. ..$ cs_coverage_0.95: num [1:24] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. .. ..$ cs_coverage_0.7 : num [1:24] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. .. ..$ cs_coverage_0.5 : num [1:24] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..$ twas_weights              :List of 6\n",
      "  .. .. ..$ enet_weights   : num [1:8057, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8057] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_weights  : num [1:8057, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8057] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_weights: num [1:8057, 1] 2.50e-04 -2.81e-04 -2.34e-04 3.92e-06 -1.15e-04 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8057] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_weights: num [1:8057, 1] 1.63e-04 -2.72e-04 -4.83e-04 -3.56e-05 -3.46e-04 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8057] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_weights  : num [1:8057, 1] 1.01e-04 -1.79e-04 -2.15e-04 -2.49e-06 -1.05e-04 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8057] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_weights  : num [1:8057, 1] 2.97e-06 -3.70e-07 -4.63e-07 -4.55e-08 -2.51e-07 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:8057] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_predictions          :List of 6\n",
      "  .. .. ..$ enet_predicted   : num [1:226, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:226] \"SM-CJEGE\" \"SM-CJEGI\" \"SM-CJEH6\" \"SM-CJEHR\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ lasso_predicted  : num [1:226, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:226] \"SM-CJEGE\" \"SM-CJEGI\" \"SM-CJEH6\" \"SM-CJEHR\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_r_predicted: num [1:226, 1] -0.1201 -0.0992 0.0867 -0.1518 -0.1329 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:226] \"SM-CJEGE\" \"SM-CJEGI\" \"SM-CJEH6\" \"SM-CJEHR\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ bayes_l_predicted: num [1:226, 1] -0.1668 -0.1163 0.0739 -0.193 -0.1576 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:226] \"SM-CJEGE\" \"SM-CJEGI\" \"SM-CJEH6\" \"SM-CJEHR\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ mrash_predicted  : num [1:226, 1] -0.0617 -0.0449 0.0364 -0.0765 -0.0598 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:226] \"SM-CJEGE\" \"SM-CJEGI\" \"SM-CJEH6\" \"SM-CJEHR\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ susie_predicted  : num [1:226, 1] -0.000317 -0.00036 0.000604 -0.000642 -0.000336 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:226] \"SM-CJEGE\" \"SM-CJEGI\" \"SM-CJEH6\" \"SM-CJEHR\" ...\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. ..$ twas_cv_result            :List of 4\n",
      "  .. .. ..$ sample_partition:'data.frame':\t226 obs. of  2 variables:\n",
      "  .. .. .. ..$ Sample: chr [1:226] \"SM-CJFK7\" \"SM-CJFKV\" \"SM-CTEDK\" \"SM-CJFLD\" ...\n",
      "  .. .. .. ..$ Fold  : int [1:226] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. ..$ prediction      :List of 6\n",
      "  .. .. .. ..$ enet_predicted   : num [1:226, 1] -0.0321 0 0 -0.016 0 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:226] \"SM-CJEGE\" \"SM-CJEGI\" \"SM-CJEH6\" \"SM-CJEHR\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ lasso_predicted  : num [1:226, 1] -0.0553 -0.02 0 -0.0241 -0.0226 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:226] \"SM-CJEGE\" \"SM-CJEGI\" \"SM-CJEH6\" \"SM-CJEHR\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_r_predicted: num [1:226, 1] -0.17975 -0.16238 -0.00551 -0.03112 -0.145 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:226] \"SM-CJEGE\" \"SM-CJEGI\" \"SM-CJEH6\" \"SM-CJEHR\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ bayes_l_predicted: num [1:226, 1] -0.1791 -0.3159 0.0169 -0.0437 -0.2801 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:226] \"SM-CJEGE\" \"SM-CJEGI\" \"SM-CJEH6\" \"SM-CJEHR\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ mrash_predicted  : num [1:226, 1] -0.08565 -0.13805 0.00143 -0.02043 -0.11234 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:226] \"SM-CJEGE\" \"SM-CJEGI\" \"SM-CJEH6\" \"SM-CJEHR\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. ..$ susie_predicted  : num [1:226, 1] 0 -0.000978 0 0 -0.000822 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:226] \"SM-CJEGE\" \"SM-CJEGI\" \"SM-CJEH6\" \"SM-CJEHR\" ...\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. ..$ performance     :List of 6\n",
      "  .. .. .. ..$ enet_performance   : num [1, 1:6] -0.04879 0.00238 -0.00207 0.46549 0.25926 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ lasso_performance  : num [1, 1:6] -0.070604 0.004985 0.000543 0.290583 0.260613 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_r_performance: num [1, 1:6] -0.013651 0.000186 -0.004277 0.838286 0.276806 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ bayes_l_performance: num [1, 1:6] -0.012197 0.000149 -0.004315 0.855311 0.292839 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ mrash_performance  : num [1, 1:6] -0.026057 0.000679 -0.003782 0.696816 0.266519 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. .. ..$ susie_performance  : num [1, 1:6] 0.025362 0.000643 -0.003818 0.704525 0.259464 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..$ : chr [1:6] \"corr\" \"rsq\" \"adj_rsq\" \"pval\" ...\n",
      "  .. .. ..$ time_elapsed    : 'proc_time' Named num [1:5] 3.405 0.886 97.332 388.485 10.095\n",
      "  .. .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ total_time_elapsed        : 'proc_time' Named num [1:5] 60.928 0.945 154.328 388.486 10.098\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:5] \"user.self\" \"sys.self\" \"elapsed\" \"user.child\" ...\n",
      "  .. ..$ variant_names             : chr [1:8057] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ region_info               :List of 3\n",
      "  .. .. ..$ region_coord:'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 64284664\n",
      "  .. .. .. ..$ end  : int 64284665\n",
      "  .. .. ..$ grange      :'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. .. ..$ chrom: chr \"11\"\n",
      "  .. .. .. ..$ start: int 61200000\n",
      "  .. .. .. ..$ end  : int 65680000\n",
      "  .. .. ..$ region_name : chr [1:11] \"ENSG00000002330\" \"ENSG00000002330\" \"ENSG00000002330\" \"chr11_ENSG00000002330_Q92934\" ...\n"
     ]
    }
   ],
   "source": [
    "# change this kernel to R code\n",
    "# TWAS weights structure\n",
    "setwd('/home/ubuntu/xqtl_protocol_exercise')\n",
    "twas_weight = readRDS('data/twas/ROSMAP_DeJager.chr11_ENSG00000002330.univariate_twas_weights.rds')\n",
    "str(twas_weight)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8abd3a22",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 6\n",
      " $ enet_weights   : num [1:7801, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. ..$ : chr [1:7801] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ : NULL\n",
      " $ lasso_weights  : num [1:7801, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. ..$ : chr [1:7801] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ : NULL\n",
      " $ bayes_r_weights: num [1:7801, 1] -1.69e-05 -1.04e-04 -1.65e-04 -1.48e-04 -9.45e-05 ...\n",
      "  ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. ..$ : chr [1:7801] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ : NULL\n",
      " $ bayes_l_weights: num [1:7801, 1] 8.28e-05 -4.19e-04 -4.27e-04 -6.72e-04 -5.37e-04 ...\n",
      "  ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. ..$ : chr [1:7801] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ : NULL\n",
      " $ mrash_weights  : num [1:7801, 1] -3.57e-06 -6.87e-05 -5.57e-05 -9.87e-05 -5.98e-05 ...\n",
      "  ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. ..$ : chr [1:7801] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ : NULL\n",
      " $ susie_weights  : num [1:7801, 1] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. ..$ : chr [1:7801] \"chr11:61201723:C:T\" \"chr11:61202327:A:G\" \"chr11:61202458:A:G\" \"chr11:61202571:G:C\" ...\n",
      "  .. ..$ : NULL\n"
     ]
    }
   ],
   "source": [
    "# r code\n",
    "twas_weight_by_method = twas_weight$ENSG00000002330$AC_DeJager_eQTL_ENSG00000002330$twas_weights\n",
    "str(twas_weight_by_method)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6869ba0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/al4225/.pixi/envs/python/lib/python3.12/site-packages/sos/targets.py:22: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n",
      "INFO: Note: NumExpr detected 32 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 16.\n",
      "INFO: NumExpr defaulting to 16 threads.\n",
      "INFO: Running \u001b[32mget_analysis_regions\u001b[0m: \n",
      "Skipping 1360 out of 1361 regions, no overlapping xQTL weights found. \n",
      "INFO: \u001b[32mget_analysis_regions\u001b[0m is \u001b[32mcompleted\u001b[0m.\n",
      "INFO: \u001b[32mget_analysis_regions\u001b[0m output:   \u001b[32mfiltered_regional_xqtl_files\u001b[0m\n",
      "INFO: Running \u001b[32mtwas\u001b[0m: \n",
      "INFO: \u001b[32mtwas\u001b[0m is \u001b[32mcompleted\u001b[0m.\n",
      "INFO: \u001b[32mtwas\u001b[0m output:   \u001b[32m/mnt/vast/hpc/homes/al4225/xqtl_protocol_data/output/twas/twas/test.chr11_84267999_86714492.twas.tsv.gz /mnt/vast/hpc/homes/al4225/xqtl_protocol_data/output/twas/twas/test.chr11_84267999_86714492.twas_data.rds... (3 items)\u001b[0m\n",
      "INFO: Workflow twas (ID=w7d134bd0a65af0da) is executed successfully with 2 completed steps.\n"
     ]
    }
   ],
   "source": [
    "cd /home/ubuntu/xqtl_protocol_exercise\n",
    "sos run pipeline/twas_ctwas.ipynb twas \\\n",
    "   --cwd output/twas --name test \\\n",
    "   --gwas_meta_data data/twas/gwas_meta_test.tsv \\\n",
    "   --ld_meta_data reference_data/ADSP_R4_EUR/ld_meta_file.tsv \\\n",
    "   --regions data/twas/EUR_LD_blocks.bed \\\n",
    "   --xqtl_meta_data data/twas/mwe_twas_pipeline_test_small.tsv \\\n",
    "   --xqtl_type_table data/twas/data_type_table.txt \\\n",
    "   --rsq_pval_cutoff 0.05 --rsq_cutoff 0.01    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39c3253f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr\tmolecular_id\tTSS\tstart\tend\tcontext\tgwas_study\tmethod\tis_imputable\tis_selected_method\trsq_cv\tpval_cv\ttwas_z\ttwas_pval\ttype\tblock\n",
      "11\tENSG00000073921\t86069881\t84957175\t87360000\tAC_DeJager_eQTL\tBellenguez_2022\tbayes_l\tTRUE\tFALSE\t0.00994563755691413\t0.0151210816406667\t-3.01085890276408\t0.00260509875533349\teQTL\tchr11_84267999_86714492\n",
      "11\tENSG00000073921\t86069881\t84957175\t87360000\tAC_DeJager_eQTL\tBellenguez_2022\tbayes_r\tTRUE\tFALSE\t0.0129222135582411\t0.00558266481673196\t-3.60984545831109\t0.000306379488103016\teQTL\tchr11_84267999_86714492\n",
      "11\tENSG00000073921\t86069881\t84957175\t87360000\tAC_DeJager_eQTL\tBellenguez_2022\tenet\tTRUE\tTRUE\t0.0302273357404713\t2.06909070605428e-05\t-3.66968282001104\t0.000242851605595364\teQTL\tchr11_84267999_86714492\n"
     ]
    }
   ],
   "source": [
    "## twas results\n",
    "cd /home/ubuntu/xqtl_protocol_exercise\n",
    "zcat output/twas/twas/test.chr11_84267999_86714492.twas.tsv.gz | head -n 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27ee370d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_name\tnum_CS\tnum_IV\tcpip\tmeta_eff\tse_meta_eff\tmeta_pval\tQ\tQ_pval\tI2\tcontext\tgwas_study\n",
      "ENSG00000073921\t\t\t\t\t\t\t\t\t\tAC_DeJager_eQTL\tBellenguez_2022\n",
      "ENSG00000073921\t\t\t\t\t\t\t\t\t\tAst_DeJager_eQTL\tBellenguez_2022\n",
      "ENSG00000073921\t\t\t\t\t\t\t\t\t\tDLPFC_Bennett_pQTL_chr11_Q13492\tBellenguez_2022\n",
      "ENSG00000073921\t\t\t\t\t\t\t\t\t\tDLPFC_DeJager_eQTL\tBellenguez_2022\n",
      "ENSG00000073921\t2\t113\t0.951\t-0.002\t0.001\t0\t1.432\t0.231\t0.302\tExc_DeJager_eQTL\tBellenguez_2022\n",
      "ENSG00000073921\t\t\t\t\t\t\t\t\t\tInh_DeJager_eQTL\tBellenguez_2022\n",
      "ENSG00000073921\t2\t8\t0.954\t-0.002\t0\t0\t0.132\t0.717\t0\tMic_DeJager_eQTL\tBellenguez_2022\n",
      "ENSG00000073921\t\t\t\t\t\t\t\t\t\tOPC_DeJager_eQTL\tBellenguez_2022\n",
      "ENSG00000073921\t1\t4\t0.985\t0.002\t0\t0\t0\t0\t0\tOli_DeJager_eQTL\tBellenguez_2022\n"
     ]
    }
   ],
   "source": [
    "## mr results\n",
    "zcat output/twas/twas/test.chr11_84267999_86714492.mr_result.tsv.gz | head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71536dbb",
   "metadata": {},
   "source": [
    "## 2. coloc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be214cb5",
   "metadata": {},
   "source": [
    "### 2.1 global enrichment analysis for the calculation of enrichment parameters between xQTL and GWAS data\n",
    "This command executes theÂ **`xqtl_gwas_enrichment`**Â workflow, which analyzes enrichment relationships between xQTL and GWAS data\n",
    "\n",
    "#### **Purpose**\n",
    "\n",
    "The purpose of this step is to calculate enrichment parameters (a0, a1) between xQTL and GWAS data and generate prior probabilities (p1, p2, p12) for subsequent colocalization analysi\n",
    "\n",
    "#### **Method**\n",
    "\n",
    "The workflow uses theÂ **`xqtl_enrichment_wrapper`**Â function to perform enrichment analysisÂ . The method:\n",
    "\n",
    "1. Identifies GWAS blocks with top loci using single variant regression methods\n",
    "2. Maps analysis regions to overlapping gene regions with corresponding QTL files containing top loci tables\n",
    "3. Finds contexts within xQTL metadata that include top loci results for each gene\n",
    "4. Executes enrichment analysis to generate parametersÂ SuSiE_enloc.ipynb:102-106\n",
    "\n",
    "#### **Input Data**\n",
    "\n",
    "**Metadata files:**\n",
    "\n",
    "- **`-gwas-meta-data`**: Metadata file for GWAS fine-mapping results\n",
    "- **`-xqtl-meta-data`**: Metadata file for xQTL fine-mapping results\n",
    "- **`-context_meta`**: Meta file showing analysis names and contained contexts\n",
    "\n",
    "**Data paths:**\n",
    "\n",
    "- **`-qtl-path`**Â andÂ **`-gwas_path`**: Directory paths for original fine-mapping results\n",
    "\n",
    "#### **Parameter Interpretation**\n",
    "\n",
    "**Object access parameters:**\n",
    "\n",
    "- **`-xqtl_finemapping_obj`**: Table name in xQTL RDS files to get fine-mapping results\n",
    "- **`-xqtl_varname_obj`**: Table name to get variable names\n",
    "- **`-gwas_finemapping_obj`**Â andÂ **`-gwas_varname_obj`**: Corresponding parameters for GWAS data\n",
    "- **`-xqtl_region_obj`**: Table name to get region information\n",
    "\n",
    "#### **Output**\n",
    "\n",
    "Generates enrichment analysis result files:Â **`*.enrichment.rds`**, containing global enrichment estimates that combine all input data for each context. The output file path format is:Â **`{cwd}/{name}.{context}.enrichment.rds`**Â \n",
    "\n",
    "#### **Relationship to Downstream Analysis**\n",
    "\n",
    "This enrichment analysis is a prerequisite step for colocalization analysis. The generated enrichment parameters serve as prior probability inputs for theÂ **`susie_coloc`**Â workflow. In colocalization analysis, the system automatically reads enrichment result files to set p1, p2, and p12 prior probabilities.\n",
    "\n",
    "#### **Notes**\n",
    "\n",
    "This workflow is the first stage of the SuSiE-enloc framework, specifically handling enrichment analysis between fine-mapping results from different genomic regions. The workflow automatically handles region overlap identification and context matching, providing statistically sound prior probabilities for subsequent colocalization analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3867e8e8",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/al4225/.pixi/envs/python/lib/python3.12/site-packages/sos/targets.py:22: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n",
      "INFO: Note: NumExpr detected 32 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 16.\n",
      "INFO: NumExpr defaulting to 16 threads.\n",
      "INFO: Running \u001b[32mget_analysis_regions\u001b[0m: get overlapped regions by gene and block region for enrichment analysis\n",
      "INFO: \u001b[32mget_analysis_regions\u001b[0m is \u001b[32mcompleted\u001b[0m.\n",
      "INFO: \u001b[32mget_analysis_regions\u001b[0m output:   \u001b[32mregional_data\u001b[0m\n",
      "INFO: Running \u001b[32mxqtl_gwas_enrichment\u001b[0m: \n",
      "INFO: \u001b[32mxqtl_gwas_enrichment\u001b[0m is \u001b[32mcompleted\u001b[0m.\n",
      "INFO: \u001b[32mxqtl_gwas_enrichment\u001b[0m output:   \u001b[32m/mnt/vast/hpc/homes/al4225/xqtl_protocol_data/output/xqtl_gwas_enrichment/demo_overlap.overlapped.demo_gwas.Knight_eQTL_brain.enrichment.rds\u001b[0m\n",
      "INFO: Workflow xqtl_gwas_enrichment (ID=w163a9b2b63d2f227) is executed successfully with 2 completed steps.\n"
     ]
    }
   ],
   "source": [
    "sos run pipeline/SuSiE_enloc.ipynb xqtl_gwas_enrichment    \\\n",
    "    --gwas-meta-data data/susie_enloc_data/demo_gwas.block_results_db.tsv \\\n",
    "    --xqtl-meta-data data/susie_enloc_data/demo_overlap.overlapped.gwas.tsv \\\n",
    "    --xqtl_finemapping_obj preset_variants_result susie_result_trimmed              \\\n",
    "    --xqtl_varname_obj preset_variants_result variant_names             \\\n",
    "    --gwas_finemapping_obj AD_Bellenguez_2022 RSS_QC_RAISS_imputed susie_result_trimmed              \\\n",
    "    --gwas_varname_obj  AD_Bellenguez_2022 RSS_QC_RAISS_imputed variant_names             \\\n",
    "    --xqtl_region_obj  region_info   grange       \\\n",
    "    --qtl-path data/susie_enloc_data \\\n",
    "    --gwas_path data/susie_enloc_data \\\n",
    "    --context_meta data/susie_enloc_data/context_meta.tsv \\\n",
    "    --cwd output/xqtl_gwas_enrichment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98507247",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 2\n",
      " $                     :List of 10\n",
      "  ..$ Alternative (coloc) p1   : num 0.000558\n",
      "  ..$ Alternative (coloc) p12  : num 6.31e-08\n",
      "  ..$ Alternative (coloc) p2   : num 0.000113\n",
      "  ..$ Effective MI rounds      : num 24\n",
      "  ..$ Enrichment (no shrinkage): num -0.409\n",
      "  ..$ Enrichment (w/ shrinkage): num -0.000151\n",
      "  ..$ Intercept                : num -7.49\n",
      "  ..$ sd (intercept)           : num 0.316\n",
      "  ..$ sd (no shrinkage)        : num 52\n",
      "  ..$ sd (w/ shrinkage)        : num 1\n",
      " $ unused_xqtl_variants:List of 1\n",
      "  ..$ : chr [1:18] \"1:20568238:C:T\" \"1:20621145:G:A\" \"1:21427833:C:G\" \"1:21688743:C:T\" ...\n"
     ]
    }
   ],
   "source": [
    "# ouput of enrichment parameter between xqtl and gwas\n",
    "# R COED\n",
    "setwd('/home/ubuntu/xqtl_protocol_exercise')\n",
    "coloc_enrichment = readRDS('output/xqtl_gwas_enrichment/demo_overlap.overlapped.demo_gwas.Knight_eQTL_brain.enrichment.rds')\n",
    "str(coloc_enrichment)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd14499a",
   "metadata": {},
   "source": [
    "### 2.2 pairwise colocalization analysis to identify shared causal variants between xQTL and GWAS\n",
    "This command executes theÂ **`susie_coloc`**Â workflow, which performs colocalization analysis between xQTL and GWAS data to identify shared causal variants.\n",
    "\n",
    "#### **Purpose**\n",
    "\n",
    "The purpose of this step is to perform pairwise colocalization analysis between xQTL and GWAS fine-mapping results to determine whether observed associations share causal variants in overlapping genomic regions. This analysis identifies contexts with top loci results for each gene and appliesÂ **`susie_coloc`**Â to analyze each gene under each identified condition.\n",
    "\n",
    "#### **Method**\n",
    "\n",
    "The workflow uses theÂ **`coloc_wrapper`**Â function to perform colocalization analysis. The method:\n",
    "\n",
    "1. **Region Overlap Detection**: Identifies GWAS blocks that contain overlapping top loci variants for each gene using genomic coordinate matching\n",
    "2. **Prior Probability Setting**: Either loads enrichment-derived priors from previous analysis or uses default values whenÂ **`-skip-enrich`**Â is specified\n",
    "3. **Colocalization Analysis**: AppliesÂ **`coloc_wrapper`**Â for each xQTL-GWAS file pair with appropriate prior probabilities\n",
    "\n",
    "#### **Input Data**\n",
    "\n",
    "**Metadata files:**\n",
    "\n",
    "- **`-gwas-meta-data`**: GWAS fine-mapping results metadata\n",
    "- **`-xqtl-meta-data`**: xQTL fine-mapping results metadata with overlap information\n",
    "- **`-context_meta`**: Analysis names and context mappings\n",
    "\n",
    "**Data paths:**\n",
    "\n",
    "- **`-qtl-path`**Â andÂ **`-gwas_path`**: Directory paths for original fine-mapping results\n",
    "\n",
    "#### **Parameter Interpretation**\n",
    "\n",
    "**Object access parameters:**\n",
    "\n",
    "- **`-xqtl_finemapping_obj`**: Table name in xQTL RDS files for fine-mapping results\n",
    "- **`-xqtl_varname_obj`**: Table name for variable names\n",
    "- **`-gwas_finemapping_obj`**Â andÂ **`-gwas_varname_obj`**: Corresponding GWAS parameters\n",
    "- **`-xqtl_region_obj`**: Table name for region information\n",
    "\n",
    "**Analysis control parameters:**\n",
    "\n",
    "- **`-skip-enrich`**: Skips enrichment analysis and uses default prior probabilities (p1=1e-4, p2=1e-4, p12=5e-6)\n",
    "- **`-ld_meta_file_path`**: LD reference metadata for post-processing credible sets\n",
    "\n",
    "#### **Output**\n",
    "\n",
    "Generates colocalization result files:Â **`*.coloc.rds`**Â containing colocalization statistics and posterior probabilities for each gene-context pair. When LD metadata is provided, also outputs credible set files (**`*.coloc_res`**) with variant-level resultsÂ SuSiE_enloc.ipynb:851Â .\n",
    "\n",
    "#### **Notes**\n",
    "\n",
    "This workflow represents the second stage of the SuSiE-enloc framework, performing gene-by-gene colocalization analysis on regions identified to have overlapping variants. TheÂ **`--skip-enrich`**Â parameter allows bypassing enrichment analysis when using default priors, making the analysis faster but potentially less statistically informed. The inclusion of LD metadata enables variant-level credible set reporting for significant colocalization results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d2a19ea",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#chr\tstart\tend\tregion_id\tTSS\toriginal_data\tcombined_data\tcombined_data_sumstats\tconditions\tconditions_top_loci\tconditions_top_loci_minp\n",
      "chr1\t17351816\t20110062\tchr1_17351816_20110062\tNA\tRSS_QC_RAISS_imputed.chr1_17351816_20110062.univariate_susie_rss.rds\tdemo_gwas.chr1_17351816_20110062.cis_results_db.export.rds\tdemo_gwas.chr1_17351816_20110062.cis_results_db.export_sumstats.rds\tAD_Bellenguez_2022,AD_Jansen_2021,AD_Kunkle_Stage1_2019,AD_Wightman_Full_2021,AD_Wightman_Excluding23andMe_2021,AD_Wightman_ExcludingUKBand23andME_2021,AD_Bellenguez_EADB_2022,AD_Bellenguez_EADI_2022\tAD_Bellenguez_2022,AD_Kunkle_Stage1_2019,AD_Bellenguez_EADB_2022\tNA\n",
      "chr1\t20110062\t22020160\tchr1_20110062_22020160\tNA\tRSS_QC_RAISS_imputed.chr1_20110062_22020160.univariate_susie_rss.rds\tdemo_gwas.chr1_20110062_22020160.cis_results_db.export.rds\tdemo_gwas.chr1_20110062_22020160.cis_results_db.export_sumstats.rds\tAD_Bellenguez_2022,AD_Jansen_2021,AD_Kunkle_Stage1_2019,AD_Wightman_Full_2021,AD_Wightman_Excluding23andMe_2021,AD_Wightman_ExcludingUKBand23andME_2021,AD_Bellenguez_EADB_2022,AD_Bellenguez_EADI_2022,AD_Bellenguez_GRACE_2022\tAD_Bellenguez_2022,AD_Jansen_2021,AD_Kunkle_Stage1_2019,AD_Wightman_Full_2021,AD_Wightman_Excluding23andMe_2021,AD_Wightman_ExcludingUKBand23andME_2021\tNA\n",
      "chr1\t22020160\t24199848\tchr1_22020160_24199848\tNA\tRSS_QC_RAISS_imputed.chr1_22020160_24199848.univariate_susie_rss.rds\tdemo_gwas.chr1_22020160_24199848.cis_results_db.export.rds\tdemo_gwas.chr1_22020160_24199848.cis_results_db.export_sumstats.rds\tAD_Bellenguez_2022,AD_Jansen_2021,AD_Kunkle_Stage1_2019,AD_Wightman_Full_2021,AD_Wightman_Excluding23andMe_2021,AD_Wightman_ExcludingUKBand23andME_2021,AD_Bellenguez_EADB_2022,AD_Bellenguez_EADI_2022,AD_Bellenguez_GRACE_2022\tAD_Bellenguez_2022,AD_Bellenguez_EADB_2022\tNA\n",
      "chr9\t19882538\t22992379\tchr9_19882538_22992379\tNA\tRSS_QC_RAISS_imputed.chr9_19882538_22992379.univariate_susie_rss.rds\tdemo_gwas.chr9_19882538_22992379.cis_results_db.export.rds\tdemo_gwas.chr9_19882538_22992379.cis_results_db.export_sumstats.rds\tAD_Bellenguez_2022,AD_Jansen_2021,AD_Kunkle_Stage1_2019,AD_Wightman_Full_2021,AD_Wightman_Excluding23andMe_2021,AD_Wightman_ExcludingUKBand23andME_2021,AD_Bellenguez_EADB_2022,AD_Bellenguez_EADI_2022,AD_Bellenguez_GRACE_2022\tAD_Kunkle_Stage1_2019,AD_Wightman_Full_2021,AD_Wightman_ExcludingUKBand23andME_2021\tNA\n"
     ]
    }
   ],
   "source": [
    "# gwas meta\n",
    "cd /home/ubuntu/xqtl_protocol_exercise\n",
    "head data/susie_enloc_data/demo_gwas.block_results_db.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c692f3c9",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/al4225/.pixi/envs/python/lib/python3.12/site-packages/sos/targets.py:22: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n",
      "INFO: Note: NumExpr detected 32 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 16.\n",
      "INFO: NumExpr defaulting to 16 threads.\n",
      "INFO: Running \u001b[32mget_overlapped_analysis_regions\u001b[0m: get overlapped regions from flatten table, to get the regions with overlapped variants and select those regions for coloc analysis\n",
      "INFO: \u001b[32mget_overlapped_analysis_regions\u001b[0m is \u001b[32mcompleted\u001b[0m.\n",
      "INFO: \u001b[32mget_overlapped_analysis_regions\u001b[0m output:   \u001b[32moverlapped_regional_data\u001b[0m\n",
      "INFO: Running \u001b[32msusie_coloc\u001b[0m: \n",
      "INFO: \u001b[32msusie_coloc\u001b[0m is \u001b[32mcompleted\u001b[0m.\n",
      "INFO: \u001b[32msusie_coloc\u001b[0m output:   \u001b[32m/mnt/vast/hpc/homes/al4225/xqtl_protocol_data/output/susie_coloc/susie_coloc/demo_overlap.overlapped.demo_gwas.Knight_eQTL_brain@ENSG00000142798.coloc.rds\u001b[0m\n",
      "INFO: Workflow susie_coloc (ID=we905e4ea186ae8c8) is executed successfully with 2 completed steps.\n"
     ]
    }
   ],
   "source": [
    "sos run pipeline/SuSiE_enloc.ipynb susie_coloc \\\n",
    "    --gwas-meta-data data/susie_enloc_data/demo_gwas.block_results_db.tsv \\\n",
    "    --xqtl-meta-data data/susie_enloc_data/demo_overlap.overlapped.gwas.tsv \\\n",
    "    --xqtl_finemapping_obj preset_variants_result susie_result_trimmed \\\n",
    "    --xqtl_varname_obj preset_variants_result variant_names \\\n",
    "    --gwas_finemapping_obj AD_Bellenguez_2022 RSS_QC_RAISS_imputed susie_result_trimmed \\\n",
    "    --gwas_varname_obj  AD_Bellenguez_2022 RSS_QC_RAISS_imputed variant_names \\\n",
    "    --xqtl_region_obj  region_info grange \\\n",
    "    --qtl-path data/susie_enloc_data \\\n",
    "    --skip-enrich \\\n",
    "    --gwas_path data/susie_enloc_data \\\n",
    "    --context_meta data/susie_enloc_data/context_meta.tsv \\\n",
    "    --ld_meta_file_path data/ld_meta_file.tsv \\\n",
    "    --cwd output/susie_coloc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7dd0e45",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 1\n",
      " $ ENSG00000142798:List of 5\n",
      "  ..$ summary        :Classes âdata.tableâ and 'data.frame':\t10 obs. of  10 variables:\n",
      "  .. ..$ nsnps    : int [1:10] 8826 8826 8826 8826 8826 8826 8826 8826 8826 8826\n",
      "  .. ..$ hit1     : chr [1:10] \"1:21912282:C:G\" \"1:21912282:C:G\" \"1:21912282:C:G\" \"1:21912282:C:G\" ...\n",
      "  .. ..$ hit2     : chr [1:10] \"1:23996556:C:T\" \"1:22187644:C:A\" \"1:22187644:C:A\" \"1:22187644:C:A\" ...\n",
      "  .. ..$ PP.H0.abf: num [1:10] 0.0165 0.0716 0.0715 0.0715 0.0715 ...\n",
      "  .. ..$ PP.H1.abf: num [1:10] 0.102 0.443 0.443 0.443 0.443 ...\n",
      "  .. ..$ PP.H2.abf: num [1:10] 0.1217 0.0645 0.0645 0.0645 0.0645 ...\n",
      "  .. ..$ PP.H3.abf: num [1:10] 0.753 0.399 0.399 0.399 0.399 ...\n",
      "  .. ..$ PP.H4.abf: num [1:10] 0.00699 0.02221 0.0222 0.0222 0.0222 ...\n",
      "  .. ..$ idx1     : int [1:10] 1 1 1 1 1 1 1 1 1 1\n",
      "  .. ..$ idx2     : int [1:10] 1 2 3 4 5 6 7 8 9 10\n",
      "  .. ..- attr(*, \".internal.selfref\")=<externalptr> \n",
      "  ..$ results        :Classes âdata.tableâ and 'data.frame':\t8826 obs. of  11 variables:\n",
      "  .. ..$ snp            : chr [1:8826] \"1:20520243:G:A\" \"1:20520537:C:A\" \"1:20521043:G:A\" \"1:20521602:T:G\" ...\n",
      "  .. ..$ SNP.PP.H4.row1 : num [1:8826] 3.28e-06 4.34e-06 4.52e-06 7.90e-06 8.21e-06 ...\n",
      "  .. ..$ SNP.PP.H4.row2 : num [1:8826] 4.49e-06 5.93e-06 6.18e-06 1.08e-05 1.12e-05 ...\n",
      "  .. ..$ SNP.PP.H4.row3 : num [1:8826] 4.49e-06 5.93e-06 6.18e-06 1.08e-05 1.12e-05 ...\n",
      "  .. ..$ SNP.PP.H4.row4 : num [1:8826] 4.49e-06 5.93e-06 6.18e-06 1.08e-05 1.12e-05 ...\n",
      "  .. ..$ SNP.PP.H4.row5 : num [1:8826] 4.49e-06 5.93e-06 6.18e-06 1.08e-05 1.12e-05 ...\n",
      "  .. ..$ SNP.PP.H4.row6 : num [1:8826] 2.74e-06 7.48e-06 5.96e-06 7.74e-06 9.18e-06 ...\n",
      "  .. ..$ SNP.PP.H4.row7 : num [1:8826] 4.26e-06 9.11e-06 7.16e-06 1.11e-05 1.30e-05 ...\n",
      "  .. ..$ SNP.PP.H4.row8 : num [1:8826] 4.26e-06 9.22e-06 7.19e-06 1.11e-05 1.30e-05 ...\n",
      "  .. ..$ SNP.PP.H4.row9 : num [1:8826] 4.26e-06 9.21e-06 7.18e-06 1.11e-05 1.30e-05 ...\n",
      "  .. ..$ SNP.PP.H4.row10: num [1:8826] 4.26e-06 9.11e-06 7.16e-06 1.11e-05 1.30e-05 ...\n",
      "  .. ..- attr(*, \".internal.selfref\")=<externalptr> \n",
      "  ..$ priors         : Named num [1:3] 1e-04 1e-04 5e-06\n",
      "  .. ..- attr(*, \"names\")= chr [1:3] \"p1\" \"p2\" \"p12\"\n",
      "  ..$ analysis_region: chr \"chr1:20520000-24080000\"\n",
      "  ..$ sets           :List of 1\n",
      "  .. ..$ cs: NULL\n"
     ]
    }
   ],
   "source": [
    "# r code\n",
    "# coloc results\n",
    "setwd('/home/ubuntu/xqtl_protocol_exercise')\n",
    "coloc_result = readRDS('output/susie_coloc/susie_coloc/demo_overlap.overlapped.demo_gwas.Knight_eQTL_brain@ENSG00000142798.coloc.rds')\n",
    "str(coloc_result)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0aab1433",
   "metadata": {},
   "source": [
    "#### Interpretation of `coloc` Analysis Results (Gene: `ENSG00000142798`)\n",
    "\n",
    "##### 1). Overall Structure\n",
    "\n",
    "The provided object is a nested list in R, structured specifically for interpreting coloc analysis results. It contains five main components:\n",
    "\n",
    "- `summary`: Posterior probabilities for different coloc hypotheses.\n",
    "- `results`: SNP-level posterior probabilities (PP.H4).\n",
    "- `priors`: Priors used in Bayesian coloc analysis.\n",
    "- `analysis_region`: Genomic region analyzed.\n",
    "- `sets`: Credible sets; here `NULL`, indicating no credible set was formed.\n",
    "\n",
    "---\n",
    "\n",
    "##### 2). `summary` Table\n",
    "\n",
    "This table contains posterior probabilities for five coloc hypotheses across SNP pairs:\n",
    "\n",
    "- `nsnps`: Number of SNPs analyzed (8826).\n",
    "- `hit1`: Representative SNP for signal 1.\n",
    "- `hit2`: Representative SNP for signal 2.\n",
    "\n",
    "Five coloc hypotheses probabilities:\n",
    "\n",
    "- `PP.H0.abf`: Neither trait has an association (null hypothesis).\n",
    "- `PP.H1.abf`: Only trait 1 is associated.\n",
    "- `PP.H2.abf`: Only trait 2 is associated.\n",
    "- `PP.H3.abf`: Both traits are associated, but at different SNPs.\n",
    "- `PP.H4.abf`: Both traits are associated at the same SNP (colocalization).\n",
    "\n",
    "Higher `PP.H4.abf` indicates stronger evidence for colocalization.\n",
    "\n",
    "---\n",
    "\n",
    "##### 3). `results` Table\n",
    "\n",
    "Contains posterior probabilities for each of the 8826 SNPs across ten coloc scenarios (`rows`):\n",
    "\n",
    "- `snp`: SNP identifier (e.g., `\"1:20520243:G:A\"`).\n",
    "- `SNP.PP.H4.row1` to `SNP.PP.H4.row10`: SNP-specific posterior probabilities for colocalization scenarios corresponding to each row in `summary`.\n",
    "\n",
    "High SNP-level posterior probabilities indicate likely colocalization positions.\n",
    "\n",
    "---\n",
    "\n",
    "##### 4). `priors`\n",
    "\n",
    "Prior probabilities used in Bayesian analysis:\n",
    "\n",
    "- `p1`: Prior probability that trait 1 is associated.\n",
    "- `p2`: Prior probability that trait 2 is associated.\n",
    "- `p12`: Prior probability of colocalization between trait 1 and trait 2.\n",
    "\n",
    "Typically set low, reflecting conservative assumptions.\n",
    "\n",
    "---\n",
    "\n",
    "##### 5). `analysis_region`\n",
    "\n",
    "Genomic region analyzed (e.g., `\"chr1:20520000-24080000\"`).\n",
    "\n",
    "---\n",
    "\n",
    "##### 6). `sets`\n",
    "\n",
    "Contains credible sets (`cs`):\n",
    "\n",
    "- Here `cs` is `NULL`, indicating no credible set formed, possibly due to weak evidence.\n",
    "\n",
    "---\n",
    "\n",
    "##### 7). Guidelines for Interpretation\n",
    "\n",
    "Interpretation of `PP.H4.abf`:\n",
    "\n",
    "- `PP.H4.abf > 0.75`: Strong evidence of colocalization.\n",
    "- `0.25 â¤ PP.H4.abf â¤ 0.75`: Moderate evidence of colocalization.\n",
    "- `PP.H4.abf < 0.25`: Weak evidence of colocalization.\n",
    "\n",
    "When `PP.H4.abf` is high, further examine individual SNP-level probabilities (`results`) to pinpoint the SNP responsible for the shared signal.\n",
    "\n",
    "---\n",
    "\n",
    "##### 8). Practical Example\n",
    "\n",
    "For row 2 of the `summary`:\n",
    "\n",
    "- `hit1`: `1:21912282:C:G`\n",
    "- `hit2`: `1:22187644:C:A`\n",
    "- `PP.H4.abf`: 0.02221 (~2.2%)\n",
    "\n",
    "The low `PP.H4.abf` indicates weak colocalization evidence. The higher `PP.H3.abf` (~40%) implies distinct SNP associations.\n",
    "\n",
    "Inspect `results` for high individual SNP posterior probabilities (e.g., > 0.9) to identify candidate colocalized SNPs.\n",
    "\n",
    "---\n",
    "\n",
    "##### 9). Summary of Key Points\n",
    "\n",
    "- Inspect `summary` for overall colocalization evidence (`PP.H4.abf`).\n",
    "- Check `results` for SNP-level posterior probabilities to identify potential colocalized SNPs.\n",
    "- A `NULL` credible set indicates insufficient evidence to confidently define a credible SNP set."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e39fad50",
   "metadata": {},
   "source": [
    "### 2.3 ColocBoost\n",
    "\n",
    "**ColocBoost** is a multi-trait colocalization analysis tool designed to identify shared genetic variants influencing multiple molecular traits.\n",
    "\n",
    "\n",
    "### Core Features\n",
    "\n",
    "- Performs colocalization analysis within genomic regions accounting for **multiple causal variants**\n",
    "- Scales to the analysis of **hundreds of traits**\n",
    "- Supports inclusion or exclusion of **GWAS summary statistics**\n",
    "- Requires **individual-level xQTL data from the same cohort**\n",
    "\n",
    "### Analysis Modes\n",
    "\n",
    "ColocBoost supports three major analysis modes:\n",
    "\n",
    "- xQTL-only Analysis (`--xqtl-coloc`):  \n",
    "  Performs colocalization only among molecular traits.\n",
    "\n",
    "- Joint GWAS Analysis (`--joint-gwas`):  \n",
    "  Integrates all GWAS traits in a combined colocalization analysis.\n",
    "\n",
    "- Separate GWAS Analysis (`--separate-gwas`):  \n",
    "  Performs colocalization analysis separately for each GWAS trait.\n",
    "\n",
    "### `--separate-gwas` vs `--no-separate-gwas`\n",
    "\n",
    "#### `--separate-gwas` (Enabled by default)\n",
    "\n",
    "When this flag is active, ColocBoost performs **individual colocalization analyses** for each GWAS trait.\n",
    "\n",
    "- Generates **one output file per GWAS study**\n",
    "- Output format: `{base_filename}.cb_xqtl_{study_name}.rds`\n",
    "- Supports **trait-specific** analysis and result inspection\n",
    "\n",
    "#### `--no-separate-gwas`\n",
    "\n",
    "When this flag is used, separate GWAS analyses are **not performed**.\n",
    "\n",
    "- Skips individual trait analysis\n",
    "- Reduces computational burden and the number of output files\n",
    "- Typically used **in combination with** `--joint-gwas`\n",
    "\n",
    "\n",
    "### Practical Examples\n",
    "\n",
    "#### xQTL-only Analysis\n",
    "\n",
    "- Use `--no-separate-gwas` to focus solely on **colocalization between molecular traits**\n",
    "\n",
    "#### GWAS-xQTL Joint Analysis\n",
    "\n",
    "- Use `--separate-gwas` to perform **trait-specific** colocalization for each GWAS dataset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71b96f66",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/al4225/.pixi/envs/python/lib/python3.12/site-packages/sos/targets.py:22: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n",
      "INFO: Note: NumExpr detected 32 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 16.\n",
      "INFO: NumExpr defaulting to 16 threads.\n",
      "INFO: Running \u001b[32mget_rss_analysis_regions\u001b[0m: \n",
      "INFO: \u001b[32mget_rss_analysis_regions\u001b[0m is \u001b[32mcompleted\u001b[0m.\n",
      "INFO: \u001b[32mget_rss_analysis_regions\u001b[0m output:   \u001b[32mregional_rss_data\u001b[0m\n",
      "INFO: Running \u001b[32mget_analysis_regions\u001b[0m: \n",
      "Loading customized association analysis window from reference_data/TAD/TADB_enhanced_cis.bed\n",
      "INFO: \u001b[32mget_analysis_regions\u001b[0m is \u001b[32mcompleted\u001b[0m.\n",
      "INFO: \u001b[32mget_analysis_regions\u001b[0m output:   \u001b[32mregional_data\u001b[0m\n",
      "INFO: Running \u001b[32mcolocboost\u001b[0m: \n",
      "INFO: \u001b[32mcolocboost\u001b[0m is \u001b[32mcompleted\u001b[0m.\n",
      "INFO: \u001b[32mcolocboost\u001b[0m output:   \u001b[32m/mnt/vast/hpc/homes/al4225/xqtl_protocol_data/output/colocboost/colocboost/test_coloc_boost_xqtl_only.chr1_ENSG00000049246.cb_xqtl.rds\u001b[0m\n",
      "INFO: Workflow colocboost (ID=w5fa65f03ec748f44) is executed successfully with 3 completed steps.\n"
     ]
    }
   ],
   "source": [
    "# xQTL-only with --no-separate-gwas parameter\n",
    "sos run pipeline/colocboost.ipynb colocboost \\\n",
    "    --name test_coloc_boost_xqtl_only  \\\n",
    "    --cwd output/colocboost \\\n",
    "    --genoFile output/plink/wgs.merged.bed \\\n",
    "    --phenoFile output/phenotype/phenotype_by_chrom_for_cis/bulk_rnaseq.phenotype_by_chrom_files.region_list.txt \\\n",
    "    --covFile output/covariate/bulk_rnaseq_tmp_matrix.low_expression_filtered.outlier_removed.tmm.expression.covariates.wgs.merged.plink_qc.plink_qc.prune.pca.Marchenko_PC.gz \\\n",
    "    --customized-association-windows reference_data/TAD/TADB_enhanced_cis.bed \\\n",
    "    --no-separate-gwas --xqtl-coloc \\\n",
    "    --region-name ENSG00000049246 \\\n",
    "    --phenotype-names trait_A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "375ecbd9",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/al4225/.pixi/envs/python/lib/python3.12/site-packages/sos/targets.py:22: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n",
      "INFO: Note: NumExpr detected 32 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 16.\n",
      "INFO: NumExpr defaulting to 16 threads.\n",
      "INFO: Running \u001b[32mget_analysis_regions\u001b[0m: \n",
      "Loading customized association analysis window from reference_data/TAD/TADB_enhanced_cis.bed\n",
      "INFO: Running \u001b[32mget_rss_analysis_regions\u001b[0m: \n",
      "Using customized association window file: reference_data/TAD/TADB_enhanced_cis.bed\n",
      "INFO: \u001b[32mget_analysis_regions\u001b[0m is \u001b[32mcompleted\u001b[0m.\n",
      "INFO: \u001b[32mget_analysis_regions\u001b[0m output:   \u001b[32mregional_data\u001b[0m\n",
      "INFO: \u001b[32mget_rss_analysis_regions\u001b[0m is \u001b[32mcompleted\u001b[0m.\n",
      "INFO: \u001b[32mget_rss_analysis_regions\u001b[0m output:   \u001b[32mregional_rss_data\u001b[0m\n",
      "INFO: Running \u001b[32mcolocboost\u001b[0m: \n",
      "INFO: \u001b[32mcolocboost\u001b[0m is \u001b[32mcompleted\u001b[0m.\n",
      "INFO: \u001b[32mcolocboost\u001b[0m output:   \u001b[32m/mnt/vast/hpc/homes/al4225/xqtl_protocol_data/output/colocboost_gwas/colocboost/colocboost_gwas.chr1_ENSG00000049246.cb_xqtl.rds /mnt/vast/hpc/homes/al4225/xqtl_protocol_data/output/colocboost_gwas/colocboost/colocboost_gwas.chr1_ENSG00000049246.cb_xqtl_Wightman.rds\u001b[0m\n",
      "INFO: Workflow colocboost (ID=we3b842e1520a2618) is executed successfully with 3 completed steps.\n"
     ]
    }
   ],
   "source": [
    "# GWAS-xQTL integrated with --separate-gwas parameter\n",
    "sos run pipeline/colocboost.ipynb colocboost \\\n",
    "    --name colocboost_gwas  \\\n",
    "    --cwd output/colocboost_gwas \\\n",
    "    --genoFile output/plink/wgs.merged.bed \\\n",
    "    --phenoFile output/phenotype/phenotype_by_chrom_for_cis/bulk_rnaseq.phenotype_by_chrom_files.region_list.txt \\\n",
    "    --covFile output/covariate/bulk_rnaseq_tmp_matrix.low_expression_filtered.outlier_removed.tmm.expression.covariates.wgs.merged.plink_qc.plink_qc.prune.pca.Marchenko_PC.gz \\\n",
    "    --customized-association-windows reference_data/TAD/TADB_enhanced_cis.bed \\\n",
    "    --ld-meta-data data/ld_meta_file.tsv \\\n",
    "    --gwas-meta-data data/colocboost/gwas_meta.txt \\\n",
    "    --separate-gwas --xqtl-coloc \\\n",
    "    --region-name ENSG00000049246 \\\n",
    "    --phenotype-names trait_A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8320531",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 1\n",
      " $ ENSG00000049246:List of 8\n",
      "  ..$ cos_summary   : NULL\n",
      "  ..$ vcp           : NULL\n",
      "  ..$ cos_details   : NULL\n",
      "  ..$ data_info     :List of 6\n",
      "  .. ..$ n_outcomes  : int 2\n",
      "  .. ..$ n_variables : int 6872\n",
      "  .. ..$ outcome_info:'data.frame':\t2 obs. of  4 variables:\n",
      "  .. .. ..$ outcome_names: chr [1:2] \"trait_A_ENSG00000049246\" \"Wightman\"\n",
      "  .. .. ..$ sample_size  : num [1:2] 150 363646\n",
      "  .. .. ..$ is_sumstats  : logi [1:2] FALSE TRUE\n",
      "  .. .. ..$ is_focal     : logi [1:2] FALSE TRUE\n",
      "  .. ..$ variables   : chr [1:6872] \"chr1:6120249:G:A\" \"chr1:6120384:G:A\" \"chr1:6120435:G:A\" \"chr1:6120926:G:A\" ...\n",
      "  .. ..$ coef        :List of 2\n",
      "  .. .. ..$ trait_A_ENSG00000049246: num [1:6872] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. ..$ Wightman               : num [1:6872] -1.30e-05 -1.15e-05 -1.18e-05 1.10e-05 -1.31e-05 ...\n",
      "  .. ..$ z           :List of 2\n",
      "  .. .. ..$ trait_A_ENSG00000049246: num [1:6872] 0 0 0 0 0 0 0 0 0 0 ...\n",
      "  .. .. ..$ Wightman               : num [1:6872] -1.0368 -0.2813 -0.4772 0.0669 -1.1019 ...\n",
      "  ..$ model_info    :List of 9\n",
      "  .. ..$ model_coveraged                : logi TRUE\n",
      "  .. ..$ n_updates                      : int 520\n",
      "  .. ..$ profile_loglik                 : num [1:520] 2 2 2 2 2 ...\n",
      "  .. ..$ outcome_profile_loglik         :List of 2\n",
      "  .. .. ..$ trait_A_ENSG00000049246: num [1:20] 1 0.989 0.984 0.98 0.976 ...\n",
      "  .. .. ..$ Wightman               : num [1:501] 1 1 1 1 0.999 ...\n",
      "  .. ..$ outcome_proximity_obj          :List of 2\n",
      "  .. .. ..$ trait_A_ENSG00000049246: num [1:20] 0.181 0.168 0.156 0.144 0.132 ...\n",
      "  .. .. ..$ Wightman               : num [1:501] 0.0106 0.0137 0.0139 0.0108 0.0116 ...\n",
      "  .. ..$ outcome_coupled_best_update_obj:List of 2\n",
      "  .. .. ..$ trait_A_ENSG00000049246: num [1:20] 0.233 0.223 0.213 0.203 0.193 ...\n",
      "  .. .. ..$ Wightman               : num [1:501] 0.0106 0.0137 0.0139 0.0108 0.0117 ...\n",
      "  .. ..$ outcome_model_coveraged        :List of 2\n",
      "  .. .. ..$ trait_A_ENSG00000049246: logi TRUE\n",
      "  .. .. ..$ Wightman               : logi FALSE\n",
      "  .. ..$ outcome_n_updates              : Named int [1:2] 20 501\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:2] \"trait_A_ENSG00000049246\" \"Wightman\"\n",
      "  .. ..$ jk_star                        : int [1:519, 1:2] NA NA NA NA NA NA NA NA NA NA ...\n",
      "  .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. ..$ : chr [1:519] \"jk_star_1\" \"jk_star_2\" \"jk_star_3\" \"jk_star_4\" ...\n",
      "  .. .. .. ..$ : chr [1:2] \"trait_A_ENSG00000049246\" \"Wightman\"\n",
      "  ..$ ucos_details  :List of 7\n",
      "  .. ..$ ucos               :List of 2\n",
      "  .. .. ..$ ucos_index    :List of 11\n",
      "  .. .. .. ..$ ucos1:y2 : int 6616\n",
      "  .. .. .. ..$ ucos2:y2 : int 6625\n",
      "  .. .. .. ..$ ucos3:y2 : int 6662\n",
      "  .. .. .. ..$ ucos4:y2 : int 6595\n",
      "  .. .. .. ..$ ucos5:y2 : int 4916\n",
      "  .. .. .. ..$ ucos6:y2 : int 6114\n",
      "  .. .. .. ..$ ucos7:y2 : int 3333\n",
      "  .. .. .. ..$ ucos8:y2 : int 2445\n",
      "  .. .. .. ..$ ucos9:y2 : int 6663\n",
      "  .. .. .. ..$ ucos10:y2: int [1:2] 6498 6617\n",
      "  .. .. .. ..$ ucos11:y2: int [1:7] 3343 3339 3344 3334 3328 3206 3338\n",
      "  .. .. ..$ ucos_variables:List of 11\n",
      "  .. .. .. ..$ ucos1:y2 : chr \"chr1:9797684:A:C\"\n",
      "  .. .. .. ..$ ucos2:y2 : chr \"chr1:9809863:C:T\"\n",
      "  .. .. .. ..$ ucos3:y2 : chr \"chr1:9830837:C:T\"\n",
      "  .. .. .. ..$ ucos4:y2 : chr \"chr1:9786186:TCAAA:T\"\n",
      "  .. .. .. ..$ ucos5:y2 : chr \"chr1:8943989:G:A\"\n",
      "  .. .. .. ..$ ucos6:y2 : chr \"chr1:9454787:T:C\"\n",
      "  .. .. .. ..$ ucos7:y2 : chr \"chr1:7900691:G:A\"\n",
      "  .. .. .. ..$ ucos8:y2 : chr \"chr1:7440907:A:G\"\n",
      "  .. .. .. ..$ ucos9:y2 : chr \"chr1:9831157:T:C\"\n",
      "  .. .. .. ..$ ucos10:y2: chr [1:2] \"chr1:9719443:A:T\" \"chr1:9799670:C:T\"\n",
      "  .. .. .. ..$ ucos11:y2: chr [1:7] \"chr1:7902251:T:C\" \"chr1:7901439:A:C\" \"chr1:7902894:T:G\" \"chr1:7900768:G:A\" ...\n",
      "  .. ..$ ucos_outcomes      :List of 2\n",
      "  .. .. ..$ outcome_index:List of 11\n",
      "  .. .. .. ..$ ucos1:y2 : num 2\n",
      "  .. .. .. ..$ ucos2:y2 : num 2\n",
      "  .. .. .. ..$ ucos3:y2 : num 2\n",
      "  .. .. .. ..$ ucos4:y2 : num 2\n",
      "  .. .. .. ..$ ucos5:y2 : num 2\n",
      "  .. .. .. ..$ ucos6:y2 : num 2\n",
      "  .. .. .. ..$ ucos7:y2 : num 2\n",
      "  .. .. .. ..$ ucos8:y2 : num 2\n",
      "  .. .. .. ..$ ucos9:y2 : num 2\n",
      "  .. .. .. ..$ ucos10:y2: num 2\n",
      "  .. .. .. ..$ ucos11:y2: num 2\n",
      "  .. .. ..$ outcome_name :List of 11\n",
      "  .. .. .. ..$ ucos1:y2 : chr \"Wightman\"\n",
      "  .. .. .. ..$ ucos2:y2 : chr \"Wightman\"\n",
      "  .. .. .. ..$ ucos3:y2 : chr \"Wightman\"\n",
      "  .. .. .. ..$ ucos4:y2 : chr \"Wightman\"\n",
      "  .. .. .. ..$ ucos5:y2 : chr \"Wightman\"\n",
      "  .. .. .. ..$ ucos6:y2 : chr \"Wightman\"\n",
      "  .. .. .. ..$ ucos7:y2 : chr \"Wightman\"\n",
      "  .. .. .. ..$ ucos8:y2 : chr \"Wightman\"\n",
      "  .. .. .. ..$ ucos9:y2 : chr \"Wightman\"\n",
      "  .. .. .. ..$ ucos10:y2: chr \"Wightman\"\n",
      "  .. .. .. ..$ ucos11:y2: chr \"Wightman\"\n",
      "  .. ..$ ucos_weight        :List of 11\n",
      "  .. .. ..$ ucos1:y2 : num [1:6872, 1] 3.60e-07 3.18e-07 3.28e-07 3.07e-07 3.64e-07 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr \"sets:Y2:ucos1\"\n",
      "  .. .. ..$ ucos2:y2 : num [1:6872, 1] 2.60e-10 2.29e-10 2.37e-10 2.21e-10 2.63e-10 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr \"sets:Y2:ucos2\"\n",
      "  .. .. ..$ ucos3:y2 : num [1:6872, 1] 4.06e-07 3.60e-07 3.71e-07 3.47e-07 4.12e-07 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr \"sets:Y2:ucos3\"\n",
      "  .. .. ..$ ucos4:y2 : num [1:6872, 1] 5.81e-07 5.13e-07 5.30e-07 4.95e-07 5.87e-07 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr \"sets:Y2:ucos5\"\n",
      "  .. .. ..$ ucos5:y2 : num [1:6872, 1] 3.58e-06 3.16e-06 3.26e-06 3.05e-06 3.62e-06 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr \"sets:Y2:ucos6\"\n",
      "  .. .. ..$ ucos6:y2 : num [1:6872, 1] 2.02e-06 1.78e-06 1.84e-06 1.72e-06 2.04e-06 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr \"sets:Y2:ucos7\"\n",
      "  .. .. ..$ ucos7:y2 : num [1:6872, 1] 1.43e-06 1.27e-06 1.31e-06 1.22e-06 1.45e-06 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr \"sets:Y2:ucos8\"\n",
      "  .. .. ..$ ucos8:y2 : num [1:6872, 1] 3.07e-06 2.71e-06 2.80e-06 2.62e-06 3.11e-06 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr \"sets:Y2:ucos11\"\n",
      "  .. .. ..$ ucos9:y2 : num [1:6872, 1] 9.27e-13 8.20e-13 8.47e-13 7.90e-13 9.39e-13 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr \"sets:Y2:ucos12\"\n",
      "  .. .. ..$ ucos10:y2: num [1:6872, 1] 1.53e-06 1.35e-06 1.39e-06 1.30e-06 1.54e-06 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr \"sets:Y2:ucos4\"\n",
      "  .. .. ..$ ucos11:y2: num [1:6872, 1] 1.46e-06 1.29e-06 1.33e-06 1.24e-06 1.48e-06 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. ..$ : chr \"sets:Y2:ucos9\"\n",
      "  .. ..$ ucos_top_variables :'data.frame':\t11 obs. of  2 variables:\n",
      "  .. .. ..$ top_index    : int [1:11] 6433 6615 6274 1218 1633 1546 1639 1735 1629 2759 ...\n",
      "  .. .. ..$ top_variables: chr [1:11] \"chr1:9673632:G:A\" \"chr1:9796577:A:G\" \"chr1:9561330:G:C\" \"chr1:6769140:T:C\" ...\n",
      "  .. ..$ ucos_purity        :List of 3\n",
      "  .. .. ..$ min_abs_cor   : num [1:11, 1:11] 1 0.32054 0.22781 0.00218 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:11] \"ucos1:y2\" \"ucos2:y2\" \"ucos3:y2\" \"ucos4:y2\" ...\n",
      "  .. .. .. .. ..$ : chr [1:11] \"ucos1:y2\" \"ucos2:y2\" \"ucos3:y2\" \"ucos4:y2\" ...\n",
      "  .. .. ..$ max_abs_cor   : num [1:11, 1:11] 1 0.32054 0.22781 0.00218 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:11] \"ucos1:y2\" \"ucos2:y2\" \"ucos3:y2\" \"ucos4:y2\" ...\n",
      "  .. .. .. .. ..$ : chr [1:11] \"ucos1:y2\" \"ucos2:y2\" \"ucos3:y2\" \"ucos4:y2\" ...\n",
      "  .. .. ..$ median_abs_cor: num [1:11, 1:11] 1 0.32054 0.22781 0.00218 0 ...\n",
      "  .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. ..$ : chr [1:11] \"ucos1:y2\" \"ucos2:y2\" \"ucos3:y2\" \"ucos4:y2\" ...\n",
      "  .. .. .. .. ..$ : chr [1:11] \"ucos1:y2\" \"ucos2:y2\" \"ucos3:y2\" \"ucos4:y2\" ...\n",
      "  .. ..$ cos_ucos_purity    : NULL\n",
      "  .. ..$ ucos_outcomes_delta:'data.frame':\t11 obs. of  2 variables:\n",
      "  .. .. ..$ ucos_outcome: chr [1:11] \"Wightman\" \"Wightman\" \"Wightman\" \"Wightman\" ...\n",
      "  .. .. ..$ ucos_delta  : num [1:11] 0.0743 0.0272 0.0134 0.015 0.0088 ...\n",
      "  ..$ region_info   :List of 3\n",
      "  .. ..$ region_coord:'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. ..$ chrom: chr \"1\"\n",
      "  .. .. ..$ start: int 7784319\n",
      "  .. .. ..$ end  : int 7845176\n",
      "  .. ..$ grange      :'data.frame':\t1 obs. of  3 variables:\n",
      "  .. .. ..$ chrom: chr \"1\"\n",
      "  .. .. ..$ start: int 6120000\n",
      "  .. .. ..$ end  : int 9960000\n",
      "  .. ..$ region_name : chr [1:2] \"ENSG00000049246\" \"ENSG00000049246\"\n",
      "  ..$ computing_time:List of 3\n",
      "  .. ..$ Loading : 'difftime' num 3.05002245505651\n",
      "  .. .. ..- attr(*, \"units\")= chr \"mins\"\n",
      "  .. ..$ QC      : 'difftime' num 17.1338541507721\n",
      "  .. .. ..- attr(*, \"units\")= chr \"secs\"\n",
      "  .. ..$ Analysis:List of 3\n",
      "  .. .. ..$ xqtl_coloc   : 'difftime' num 14.8683335781097\n",
      "  .. .. .. ..- attr(*, \"units\")= chr \"secs\"\n",
      "  .. .. ..$ joint_gwas   : NULL\n",
      "  .. .. ..$ separate_gwas:List of 3\n",
      "  .. .. .. ..$ total    : 'difftime' num 4.68074152469635\n",
      "  .. .. .. .. ..- attr(*, \"units\")= chr \"mins\"\n",
      "  .. .. .. ..$ n_studies: int 1\n",
      "  .. .. .. ..$ average  : 'difftime' num 4.68074152469635\n",
      "  .. .. .. .. ..- attr(*, \"units\")= chr \"mins\"\n",
      "  ..- attr(*, \"class\")= chr \"colocboost\"\n"
     ]
    }
   ],
   "source": [
    "# output of colocboost\n",
    "setwd('/home/ubuntu/xqtl_protocol_exercise')\n",
    "colocboost = readRDS('output/colocboost_gwas/colocboost/colocboost_gwas.chr1_ENSG00000049246.cb_xqtl_Wightman.rds')\n",
    "str(colocboost)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bcfbfad",
   "metadata": {},
   "source": [
    "\n",
    "### Colocboost Output File Structure\n",
    "\n",
    "- **When `--separate-gwas` is enabled**:  \n",
    "  ColocBoost outputs **one RDS file per GWAS trait**, containing:\n",
    "  - Region-level colocalization results\n",
    "  - Metadata including analysis time, number of SNPs, etc.\n",
    "\n",
    "- File naming convention follows:  \n",
    "  `{base_filename}.cb_xqtl_{study_name}.rds`\n",
    "\n",
    "This output is generated from a ColocBoost analysis examining shared causal variants between xQTL and GWAS data.\n",
    "\n",
    "\n",
    "\n",
    "The `colocboost` result is a named list of class `\"colocboost\"`, where each top-level element corresponds to a gene (e.g., `ENSG00000049246`). The internal structure is divided into several components:\n",
    "\n",
    "---\n",
    "\n",
    "#### Top-Level Components\n",
    "\n",
    "| Field | Type | Description |\n",
    "| --- | --- | --- |\n",
    "| `cos_summary` | NULL or list | Summary of colocalized signals (empty if none) |\n",
    "| `vcp` | NULL or list | Variant colocalization probabilities |\n",
    "| `cos_details` | NULL or list | Details for colocalized signals |\n",
    "| `data_info` | list | Information on traits, SNPs, z-scores, and coefficients |\n",
    "| `model_info` | list | Model convergence and log-likelihood tracking |\n",
    "| `ucos_details` | list | Results for **uncolocalized signals** (unique to colocboost) |\n",
    "| `region_info` | list | Genomic coordinates and extended window used in analysis |\n",
    "| `computing_time` | list | Timing for loading, QC, and model fitting steps |\n",
    "\n",
    "---\n",
    "\n",
    "#### data_info\n",
    "\n",
    "| Field | Description |\n",
    "| --- | --- |\n",
    "| `n_outcomes` | Number of traits analyzed |\n",
    "| `n_variables` | Number of SNPs analyzed |\n",
    "| `outcome_info` | Data frame of each traitâs name, sample size, is_sumstats, and is_focal |\n",
    "| `variables` | Character vector of SNP IDs in `chr:pos:ref:alt` format |\n",
    "| `coef`, `z` | Trait-wise regression coefficients and z-scores (lists with length = number of traits) |\n",
    "\n",
    "---\n",
    "\n",
    "#### model_info\n",
    "\n",
    "Tracks model convergence and optimization metrics for each trait:\n",
    "\n",
    "| Field | Description |\n",
    "| --- | --- |\n",
    "| `model_coveraged` | Whether the joint model converged |\n",
    "| `outcome_model_coveraged` | Per-trait convergence status |\n",
    "| `n_updates` | Total number of updates |\n",
    "| `outcome_n_updates` | Number of updates per trait |\n",
    "| `profile_loglik` | Log-likelihood per iteration (joint) |\n",
    "| `outcome_profile_loglik` | Log-likelihood per trait |\n",
    "| `outcome_proximity_obj` | Proximity statistics per trait per iteration |\n",
    "| `outcome_coupled_best_update_obj` | Coupled update quality per trait |\n",
    "| `jk_star` | Jackknife matrix (optional) |\n",
    "\n",
    "---\n",
    "\n",
    "#### ucos_details\n",
    "\n",
    "Describes **uncolocalized effect sets** when no coloc is detected:\n",
    "\n",
    "| Field | Description |\n",
    "| --- | --- |\n",
    "| `ucos_index` | Indices of SNPs in each uncolocalized set |\n",
    "| `ucos_variables` | SNP IDs in each set |\n",
    "| `ucos_outcomes` | Trait name associated with each set (e.g., `\"Wightman\"`) |\n",
    "| `ucos_weight` | Per-SNP weights in each set |\n",
    "| `ucos_top_variables` | Top SNP per set |\n",
    "| `ucos_purity` | Pairwise correlation (min, max, median) across sets |\n",
    "| `ucos_outcomes_delta` | Delta value (importance) per set |\n",
    "\n",
    "> Interpretation:\n",
    "> \n",
    "> - Large number of `ucos` suggests GWAS has signals independent of molecular QTLs.\n",
    "> - `ucos_delta` indicates signal strength; higher = more trait variance explained.\n",
    "\n",
    "---\n",
    "\n",
    "#### region_info\n",
    "\n",
    "| Field | Description |\n",
    "| --- | --- |\n",
    "| `region_coord` | Original gene start/end coordinates |\n",
    "| `grange` | Extended region used for coloc (typically Â±1.5Mb) |\n",
    "| `region_name` | Gene or region ID (can repeat if multiple traits mapped) |\n",
    "\n",
    "#### computing_time\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9ee8781",
   "metadata": {},
   "source": [
    "\n",
    "#### Key Results Interpretation in this MWE\n",
    "\n",
    "#### Analysis Overview\n",
    "\n",
    "- **Gene**: ENSG00000049246 (chromosome 1)\n",
    "- **Traits analyzed**:  \n",
    "  - `trait_A_ENSG00000049246` (xQTL, n = 150)  \n",
    "  - `Wightman` (GWAS, n = 363,646)\n",
    "- **Variants**: 6,872 variants analyzed within the region\n",
    "\n",
    "#### Missing Core Results\n",
    "\n",
    "- `cos_summary`: `NULL` â No colocalized effects detected  \n",
    "- `vcp`: `NULL` â No variant colocalization probabilities computed  \n",
    "- `cos_details`: `NULL` â No detailed colocalization results available\n",
    "\n",
    "This indicates no significant colocalization between xQTL and GWAS signals for the target gene.\n",
    "\n",
    "These results suggest that while the Wightman GWAS shows multiple independent association signals in the region, these signals do not colocalize with the xQTL effects. \n",
    "\n",
    "### Uncolocalized Effects (`ucos_details`)\n",
    "\n",
    "The analysis detected 11 uncolocalized effects (`ucos1:y2` to `ucos11:y2`), all associated with the GWAS trait `Wightman`.  \n",
    "\n",
    "- **Top variants**:  \n",
    "  - `chr1:9797684:A:C`  \n",
    "  - `chr1:9809863:C:T`  \n",
    "  - `chr1:9830837:C:T`  \n",
    "- **Effect strengths**:  \n",
    "  - Ranged from extremely weak (`ucos9:y2` with weights â 9e-13)  \n",
    "  - To moderate (`ucos5:y2` with weights â 3e-06)  \n",
    "- **Delta values**:  \n",
    "  - Effect size contributions ranged from 0.0088 to 0.0743\n",
    "\n",
    "### Model Convergence\n",
    "\n",
    "- `trait_A_ENSG00000049246`: Converged after 20 updates (`model_coveraged = TRUE`)\n",
    "- `Wightman`: Did not converge after 501 updates (`model_coveraged = FALSE`)\n",
    "\n",
    "\n",
    "### Technical Details\n",
    "\n",
    "- **Analysis region**: chr1:6,120,000â9,960,000  \n",
    "- **Gene coordinates**: chr1:7,784,319â7,845,176  \n",
    "\n",
    "**Computing time**:  \n",
    "- Data loading: ~3 minutes  \n",
    "- Quality control: ~17 seconds  \n",
    "- Model fitting:  \n",
    "  - xQTL: ~15 seconds  \n",
    "  - GWAS integration: ~4.7 minutes"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R (pisces-rabbit)",
   "language": "R",
   "name": "r-pisces-rabbit"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
